**Sabrina Zheng | Group 6 | Chapter 5 Questions 3, 4, 5, 6**

**Question 3:**

**\-- 3-1**

\-- Write a query that calculates a row number for each order

\-- based on orderdate, orderid ordering

\-- Tables involved: Sales.Orders

\-- Desired output:

orderid     orderdate   custid      empid       rownum

\----------- ----------- ----------- ----------- -------

10248       2020-07-04  85          5           1

10249       2020-07-05  79          6           2

10250       2020-07-08  34          4           3

10251       2020-07-08  84          3           4

10252       2020-07-09  76          4           5

10253       2020-07-10  34          3           6

10254       2020-07-11  14          5           7

10255       2020-07-12  68          9           8

10256       2020-07-15  88          3           9

10257       2020-07-16  35          4           10

In [114]:
--Query using TSQLV6 database
USE TSQLV6;

SELECT 
    orderid,
    orderdate,
    custid,
    empid,
    ROW_NUMBER() OVER (ORDER BY orderdate, orderid) AS rownum
FROM 
    Sales.Orders;

(830 rows affected)

Total execution time: 00:00:00.077

orderid orderdate custid empid rownum 10248 2020-07-04 85 5 1 10249 2020-07-05 79 6 2 10250 2020-07-08 34 4 3 10251 2020-07-08 84 3 4 10252 2020-07-09 76 4 5 10253 2020-07-10 34 3 6 10254 2020-07-11 14 5 7 10255 2020-07-12 68 9 8 10256 2020-07-15 88 3 9 10257 2020-07-16 35 4 10 10258 2020-07-17 20 1 11 10259 2020-07-18 13 4 12 10260 2020-07-19 56 4 13 10261 2020-07-19 61 4 14 10262 2020-07-22 65 8 15 10263 2020-07-23 20 9 16 10264 2020-07-24 24 6 17 10265 2020-07-25 7 2 18 10266 2020-07-26 87 3 19 10267 2020-07-29 25 4 20 10268 2020-07-30 33 8 21 10269 2020-07-31 89 5 22 10270 2020-08-01 87 1 23 10271 2020-08-01 75 6 24 10272 2020-08-02 65 6 25 10273 2020-08-05 63 3 26 10274 2020-08-06 85 6 27 10275 2020-08-07 49 1 28 10276 2020-08-08 80 8 29 10277 2020-08-09 52 2 30 10278 2020-08-12 5 8 31 10279 2020-08-13 44 8 32 10280 2020-08-14 5 2 33 10281 2020-08-14 69 4 34 10282 2020-08-15 69 4 35 10283 2020-08-16 46 3 36 10284 2020-08-19 44 4 37 10285 2020-08-20 63 1 38 10286 2020-08-21 63 8 39 10287 2020-08-22 67 8 40 10288 2020-08-23 66 4 41 10289 2020-08-26 11 7 42 10290 2020-08-27 15 8 43 10291 2020-08-27 61 6 44 10292 2020-08-28 81 1 45 10293 2020-08-29 80 1 46 10294 2020-08-30 65 4 47 10295 2020-09-02 85 2 48 10296 2020-09-03 46 6 49 10297 2020-09-04 7 5 50 10298 2020-09-05 37 6 51 10299 2020-09-06 67 4 52 10300 2020-09-09 49 2 53 10301 2020-09-09 86 8 54 10302 2020-09-10 76 4 55 10303 2020-09-11 30 7 56 10304 2020-09-12 80 1 57 10305 2020-09-13 55 8 58 10306 2020-09-16 69 1 59 10307 2020-09-17 48 2 60 10308 2020-09-18 2 7 61 10309 2020-09-19 37 3 62 10310 2020-09-20 77 8 63 10311 2020-09-20 18 1 64 10312 2020-09-23 86 2 65 10313 2020-09-24 63 2 66 10314 2020-09-25 65 1 67 10315 2020-09-26 38 4 68 10316 2020-09-27 65 1 69 10317 2020-09-30 48 6 70 10318 2020-10-01 38 8 71 10319 2020-10-02 80 7 72 10320 2020-10-03 87 5 73 10321 2020-10-03 38 3 74 10322 2020-10-04 58 7 75 10323 2020-10-07 39 4 76 10324 2020-10-08 71 9 77 10325 2020-10-09 39 1 78 10326 2020-10-10 8 4 79 10327 2020-10-11 24 2 80 10328 2020-10-14 28 4 81 10329 2020-10-15 75 4 82 10330 2020-10-16 46 3 83 10331 2020-10-16 9 9 84 10332 2020-10-17 51 3 85 10333 2020-10-18 87 5 86 10334 2020-10-21 84 8 87 10335 2020-10-22 37 7 88 10336 2020-10-23 60 7 89 10337 2020-10-24 25 4 90 10338 2020-10-25 55 4 91 10339 2020-10-28 51 2 92 10340 2020-10-29 9 1 93 10341 2020-10-29 73 7 94 10342 2020-10-30 25 4 95 10343 2020-10-31 44 4 96 10344 2020-11-01 89 4 97 10345 2020-11-04 63 2 98 10346 2020-11-05 65 3 99 10347 2020-11-06 21 4 100 10348 2020-11-07 86 4 101 10349 2020-11-08 75 7 102 10350 2020-11-11 41 6 103 10351 2020-11-11 20 1 104 10352 2020-11-12 28 3 105 10353 2020-11-13 59 7 106 10354 2020-11-14 58 8 107 10355 2020-11-15 4 6 108 10356 2020-11-18 86 6 109 10357 2020-11-19 46 1 110 10358 2020-11-20 41 5 111 10359 2020-11-21 72 5 112 10360 2020-11-22 7 4 113 10361 2020-11-22 63 1 114 10362 2020-11-25 9 3 115 10363 2020-11-26 17 4 116 10364 2020-11-26 19 1 117 10365 2020-11-27 3 3 118 10366 2020-11-28 29 8 119 10367 2020-11-28 83 7 120 10368 2020-11-29 20 2 121 10369 2020-12-02 75 8 122 10370 2020-12-03 14 6 123 10371 2020-12-03 41 1 124 10372 2020-12-04 62 5 125 10373 2020-12-05 37 4 126 10374 2020-12-05 91 1 127 10375 2020-12-06 36 3 128 10376 2020-12-09 51 1 129 10377 2020-12-09 72 1 130 10378 2020-12-10 24 5 131 10379 2020-12-11 61 2 132 10380 2020-12-12 37 8 133 10381 2020-12-12 46 3 134 10382 2020-12-13 20 4 135 10383 2020-12-16 4 8 136 10384 2020-12-16 5 3 137 10385 2020-12-17 75 1 138 10386 2020-12-18 21 9 139 10387 2020-12-18 70 1 140 10388 2020-12-19 72 2 141 10389 2020-12-20 10 4 142 10390 2020-12-23 20 6 143 10391 2020-12-23 17 3 144 10392 2020-12-24 59 2 145 10393 2020-12-25 71 1 146 10394 2020-12-25 36 1 147 10395 2020-12-26 35 6 148 10396 2020-12-27 25 1 149 10397 2020-12-27 60 5 150 10398 2020-12-30 71 2 151 10399 2020-12-31 83 8 152 10400 2021-01-01 19 1 153 10401 2021-01-01 65 1 154 10402 2021-01-02 20 8 155 10403 2021-01-03 20 4 156 10404 2021-01-03 49 2 157 10

**Explanation for the TSQLV6 Query**

**Proposition:** Calculate a row number for each order based on orderid and order date using the Sales.Orders table.

**Tables:** 

- Sales.Orders

**Columns:**

- orderid
- orderdate
- custid
- empid

**Predicate:**

- Retrieves data from Sales.Orders table
- Selects the columns orderid, orderdate, custid, and empid.
- Uses the ROW\_NUMBER() window function to assign a sequential row number to each row in the result set.
- The ROW\_NUMBER() function is ordered by orderdate and orderid in ascending order.

In [2]:
-- Query using Northwinds2022TSQLV7 database 
USE Northwinds2022TSQLV7;

SELECT 
    orderid,
    orderdate,
    customerid,
    employeeid,
    ROW_NUMBER() OVER (ORDER BY orderdate, orderid) AS rownum
FROM 
    Sales.[Order];

(830 rows affected)

Total execution time: 00:00:00.024

orderid orderdate customerid employeeid rownum 10248 2014-07-04 85 5 1 10249 2014-07-05 79 6 2 10250 2014-07-08 34 4 3 10251 2014-07-08 84 3 4 10252 2014-07-09 76 4 5 10253 2014-07-10 34 3 6 10254 2014-07-11 14 5 7 10255 2014-07-12 68 9 8 10256 2014-07-15 88 3 9 10257 2014-07-16 35 4 10 10258 2014-07-17 20 1 11 10259 2014-07-18 13 4 12 10260 2014-07-19 56 4 13 10261 2014-07-19 61 4 14 10262 2014-07-22 65 8 15 10263 2014-07-23 20 9 16 10264 2014-07-24 24 6 17 10265 2014-07-25 7 2 18 10266 2014-07-26 87 3 19 10267 2014-07-29 25 4 20 10268 2014-07-30 33 8 21 10269 2014-07-31 89 5 22 10270 2014-08-01 87 1 23 10271 2014-08-01 75 6 24 10272 2014-08-02 65 6 25 10273 2014-08-05 63 3 26 10274 2014-08-06 85 6 27 10275 2014-08-07 49 1 28 10276 2014-08-08 80 8 29 10277 2014-08-09 52 2 30 10278 2014-08-12 5 8 31 10279 2014-08-13 44 8 32 10280 2014-08-14 5 2 33 10281 2014-08-14 69 4 34 10282 2014-08-15 69 4 35 10283 2014-08-16 46 3 36 10284 2014-08-19 44 4 37 10285 2014-08-20 63 1 38 10286 2014-08-21 63 8 39 10287 2014-08-22 67 8 40 10288 2014-08-23 66 4 41 10289 2014-08-26 11 7 42 10290 2014-08-27 15 8 43 10291 2014-08-27 61 6 44 10292 2014-08-28 81 1 45 10293 2014-08-29 80 1 46 10294 2014-08-30 65 4 47 10295 2014-09-02 85 2 48 10296 2014-09-03 46 6 49 10297 2014-09-04 7 5 50 10298 2014-09-05 37 6 51 10299 2014-09-06 67 4 52 10300 2014-09-09 49 2 53 10301 2014-09-09 86 8 54 10302 2014-09-10 76 4 55 10303 2014-09-11 30 7 56 10304 2014-09-12 80 1 57 10305 2014-09-13 55 8 58 10306 2014-09-16 69 1 59 10307 2014-09-17 48 2 60 10308 2014-09-18 2 7 61 10309 2014-09-19 37 3 62 10310 2014-09-20 77 8 63 10311 2014-09-20 18 1 64 10312 2014-09-23 86 2 65 10313 2014-09-24 63 2 66 10314 2014-09-25 65 1 67 10315 2014-09-26 38 4 68 10316 2014-09-27 65 1 69 10317 2014-09-30 48 6 70 10318 2014-10-01 38 8 71 10319 2014-10-02 80 7 72 10320 2014-10-03 87 5 73 10321 2014-10-03 38 3 74 10322 2014-10-04 58 7 75 10323 2014-10-07 39 4 76 10324 2014-10-08 71 9 77 10325 2014-10-09 39 1 78 10326 2014-10-10 8 4 79 10327 2014-10-11 24 2 80 10328 2014-10-14 28 4 81 10329 2014-10-15 75 4 82 10330 2014-10-16 46 3 83 10331 2014-10-16 9 9 84 10332 2014-10-17 51 3 85 10333 2014-10-18 87 5 86 10334 2014-10-21 84 8 87 10335 2014-10-22 37 7 88 10336 2014-10-23 60 7 89 10337 2014-10-24 25 4 90 10338 2014-10-25 55 4 91 10339 2014-10-28 51 2 92 10340 2014-10-29 9 1 93 10341 2014-10-29 73 7 94 10342 2014-10-30 25 4 95 10343 2014-10-31 44 4 96 10344 2014-11-01 89 4 97 10345 2014-11-04 63 2 98 10346 2014-11-05 65 3 99 10347 2014-11-06 21 4 100 10348 2014-11-07 86 4 101 10349 2014-11-08 75 7 102 10350 2014-11-11 41 6 103 10351 2014-11-11 20 1 104 10352 2014-11-12 28 3 105 10353 2014-11-13 59 7 106 10354 2014-11-14 58 8 107 10355 2014-11-15 4 6 108 10356 2014-11-18 86 6 109 10357 2014-11-19 46 1 110 10358 2014-11-20 41 5 111 10359 2014-11-21 72 5 112 10360 2014-11-22 7 4 113 10361 2014-11-22 63 1 114 10362 2014-11-25 9 3 115 10363 2014-11-26 17 4 116 10364 2014-11-26 19 1 117 10365 2014-11-27 3 3 118 10366 2014-11-28 29 8 119 10367 2014-11-28 83 7 120 10368 2014-11-29 20 2 121 10369 2014-12-02 75 8 122 10370 2014-12-03 14 6 123 10371 2014-12-03 41 1 124 10372 2014-12-04 62 5 125 10373 2014-12-05 37 4 126 10374 2014-12-05 91 1 127 10375 2014-12-06 36 3 128 10376 2014-12-09 51 1 129 10377 2014-12-09 72 1 130 10378 2014-12-10 24 5 131 10379 2014-12-11 61 2 132 10380 2014-12-12 37 8 133 10381 2014-12-12 46 3 134 10382 2014-12-13 20 4 135 10383 2014-12-16 4 8 136 10384 2014-12-16 5 3 137 10385 2014-12-17 75 1 138 10386 2014-12-18 21 9 139 10387 2014-12-18 70 1 140 10388 2014-12-19 72 2 141 10389 2014-12-20 10 4 142 10390 2014-12-23 20 6 143 10391 2014-12-23 17 3 144 10392 2014-12-24 59 2 145 10393 2014-12-25 71 1 146 10394 2014-12-25 36 1 147 10395 2014-12-26 35 6 148 10396 2014-12-27 25 1 149 10397 2014-12-27 60 5 150 10398 2014-12-30 71 2 151 10399 2014-12-31 83 8 152 10400 2015-01-01 19 1 153 10401 2015-01-01 65 1 154 10402 2015-01-02 20 8 155 10403 2015-01-03 20 4 156 10404 2015-01-03 49

### **Explanation for the Northwinds2022TSQLV7 Query**

**Proposition:** <span style="color: var(--vscode-foreground);">Calculate a row number for each order based on orderid and order date using the Sales.Orders table.</span>

**Tables:** 

- Sales.\[Order\]

**Columns:**

- orderid
- orderdate
- customerid
- emplyeeid

**Predicate:**

- ROW\_NUMBER() OVER (ORDER BY orderdate, orderid) AS rownum

 <span style="color: var(--vscode-foreground);">&nbsp;This is a window function that assigns a sequential row number to each row in the result set. The ORDER BY clause within the OVER function specifies the ordering criteria for assigning row numbers. In this case, the rows will be ordered by orderdate first and then by orderid.</span>

**\-- 3-2**

\-- Write a query that returns rows with row numbers 11 through 20

\-- based on the row number definition in exercise 3-1

\-- Use a CTE to encapsulate the code from exercise 3-1

\-- Tables involved: Sales.Orders

\-- Desired output:

orderid     orderdate   custid      empid       rownum

\----------- ----------- ----------- ----------- -------

10258       2020-07-17  20          1           11

10259       2020-07-18  13          4           12

10260       2020-07-19  56          4           13

10261       2020-07-19  61          4           14

10262       2020-07-22  65          8           15

10263       2020-07-23  20          9           16

10264       2020-07-24  24          6           17

10265       2020-07-25  7           2           18

10266       2020-07-26  87          3           19

10267       2020-07-29  25          4           20

In [115]:
--Query using TSQLV6 database
USE TSQLV6;

WITH OrderedOrders AS (
    SELECT 
        orderid,
        orderdate,
        custid AS custid,
        empid AS empid,
        ROW_NUMBER() OVER (ORDER BY orderdate, orderid) AS rownum
    FROM 
        Sales.Orders
)

SELECT 
    orderid,
    orderdate,
    custid,
    empid,
    rownum
FROM 
    OrderedOrders
WHERE 
    rownum BETWEEN 11 AND 20;

(10 rows affected)

Total execution time: 00:00:00.044

orderid orderdate custid empid rownum 10258 2020-07-17 20 1 11 10259 2020-07-18 13 4 12 10260 2020-07-19 56 4 13 10261 2020-07-19 61 4 14 10262 2020-07-22 65 8 15 10263 2020-07-23 20 9 16 10264 2020-07-24 24 6 17 10265 2020-07-25 7 2 18 10266 2020-07-26 87 3 19 10267 2020-07-29 25 4 20

**Explanation for the TSQLV6 Query**

**Proposition:** Retrieve a subset of rows from the Sales.Orders table based on the row numbers between 11 and 20 inclusive

**Tables:** 

- Sales.Orders

**Columns:**

- orderid
- orderdate
- custid
- empid
- ROW\_NUMBER() OVER (ORDER BY orderdate, orderid) AS rownum

<span style="color: var(--vscode-foreground);">The ORDER BY clause within the OVER function specifies the ordering criteria for assigning row numbers. In this case, the rows will be ordered by orderdate first and then by orderid.</span>

**Predicate:**

- WHERE rownum BETWEEN 11 AND 20

<span style="color: var(--vscode-foreground);">This&nbsp; filters the rows from the CTE (OrderedOrders) to include only those with row numbers between 11 and 20.</span>

In [4]:
-- Query using Northwinds2022TSQLV7 database 
USE Northwinds2022TSQLV7;

WITH OrderedOrders AS (
    SELECT 
        orderid,
        orderdate,
        customerid AS custid,
        employeeid AS empid,
        ROW_NUMBER() OVER (ORDER BY orderdate, orderid) AS rownum
    FROM 
        Sales.[Order]
)

SELECT 
    orderid,
    orderdate,
    custid,
    empid,
    rownum
FROM 
    OrderedOrders
WHERE 
    rownum BETWEEN 11 AND 20;

(10 rows affected)

Total execution time: 00:00:00.012

orderid orderdate custid empid rownum 10258 2014-07-17 20 1 11 10259 2014-07-18 13 4 12 10260 2014-07-19 56 4 13 10261 2014-07-19 61 4 14 10262 2014-07-22 65 8 15 10263 2014-07-23 20 9 16 10264 2014-07-24 24 6 17 10265 2014-07-25 7 2 18 10266 2014-07-26 87 3 19 10267 2014-07-29 25 4 20

### **Explanation for the Northwinds2022TSQLV7 Query**

**Tables:**

- Sales.\[Order\]

**Columns:**

- orderid
- orderdate
- customerid
- employeeid

CTE:

- OrderedOrders

**Predicate:**  

- CTE OrderedOrders is created, which selects columns orderid, orderdate, customerid (aliased as custid), employeeid (aliased as empid), and calculates the row number using the ROW\_NUMBER() window function. The rows are ordered by orderdate first and then by orderid.
- The selected columns (orderid, orderdate, custid, empid, rownum) are retrieved from the CTE OrderedOrders.
- <span style="color: var(--vscode-foreground);">A WHERE clause is applied to filter the rows where the rownum is between 11 and 20, inclusive.</span>
- The predicate in this query is rownum BETWEEN 11 AND 20, which filters the rows based on the row number calculated in the CTE. It selects rows with row numbers between 11 and 20.

**Question 4:**

\-- Write a solution using a recursive CTE that returns the 

\-- management chain leading to Patricia Doyle (employee ID 9)

\-- Tables involved: HR.Employees

\-- Desired output:

empid       mgrid       firstname  lastname

\----------- ----------- ---------- --------------------

9           5           Patricia   Doyle

5           2           Sven       Mortensen

2           1           Don        Funk

1           NULL        Sara       Davis

In [118]:
--Query using TSQLV6 database
USE TSQLV6;

WITH RecursiveHierarchy AS (
    SELECT 
        empid,
        mgrid,
        firstname,
        lastname
    FROM 
        HR.Employees
    WHERE 
        empid = 9  -- Starting point: Patricia Doyle
    UNION ALL
    SELECT 
        e.empid,
        e.mgrid,
        e.firstname,
        e.lastname
    FROM 
        HR.Employees e
    INNER JOIN 
        RecursiveHierarchy rh ON e.empid = rh.mgrid
)

SELECT 
    empid,
    mgrid,
    firstname,
    lastname
FROM 
    RecursiveHierarchy;

(4 rows affected)

Total execution time: 00:00:00.016

empid mgrid firstname lastname 9 5 Patricia Doyle 5 2 Sven Mortensen 2 1 Don Funk 1 NULL Sara Davis

**Explanation for the TSQLV6 Query**

**Proposition:** Return the management chain leading to Patricia Doyle from the HR.Employees table using a recursive CTE

**Table:**

- HR.Employees

**Columns:**

- empid
- mgrid
- firstname
- lastname

Recursive CTE:

- RecursiveHierarchy

**Predicate:**

- <span style="color: var(--vscode-foreground);">The recursive CTE RecursiveHierarchy begins by selecting Patricia Doyle (employee ID 9) as the starting point from the HR.Employees table</span>
    
- In the recursive part of the CTE, it joins the HR.Employees table with the previous result (RecursiveHierarchy) based on the mgrid column to find the manager of each employee. This continues recursively until no more managers are found.
    
- The last SELECT statement retrieves all columns (empid, mgrid, firstname, lastname) from the RecursiveHierarchy, which now contains the entire management chain leading to Patricia Doyle.
    
- The CTE (Common Table Expression) recursively navigates the management hierarchy starting from Patricia Doyle (employee ID 9) until reaching the top of the hierarchy where the manager's mgrid is NULL.

In [6]:
-- Query using Northwinds2022TSQLV7 database 
USE Northwinds2022TSQLV7;

WITH RecursiveHierarchy AS (
    SELECT 
        employeeid,
        EmployeeManagerId,
        EmployeeFirstName,
        EmployeeLastName
    FROM 
        HumanResources.[Employee]
    WHERE 
        employeeid = 9  -- Starting point: Patricia Doyle
    UNION ALL
    SELECT 
        e.employeeid,
        e.EmployeeManagerId,
        e.EmployeeFirstName,
        e.EmployeeLastName
    FROM 
        HumanResources.[Employee] e
    INNER JOIN 
        RecursiveHierarchy rh ON e.employeeid = rh.EmployeeManagerId
)

SELECT 
    employeeid,
    EmployeeManagerId,
    EmployeeFirstName,
    EmployeeLastName
FROM 
    RecursiveHierarchy;

(4 rows affected)

Total execution time: 00:00:00.025

employeeid EmployeeManagerId EmployeeFirstName EmployeeLastName 9 5 Patricia Doyle 5 2 Sven Mortensen 2 1 Don Funk 1 NULL Sara Davis

### **Explanation for the Northwinds2022TSQLV7 Query**

**Proposition:** <span style="color: var(--vscode-foreground);">Return the management chain leading to Patricia Doyle from the HR.Employees table using a recursive CTE</span>

**Tables:**

- HumanResources.\[Employee\]

HumanResources.Employee: This is the table from which data is being retrieved. It's likely a table named "Employee" in the "HumanResources" schema of the "Northwinds2022TSQLV7" database. This table presumably contains information about employees, including their IDs (employeeid), manager IDs (EmployeeManagerId), first names (EmployeeFirstName), and last names (EmployeeLastName).

**Columns:**

- emplyeeid
- EmployeeManagerId
- EmployeeFirstName
- EmployeeLastName

Recursive CTE:

- RecursiveHierarchy

**Predicate:**  

- Recursive CTE RecursiveHierarchy begins by selecting Patricia Doyle (employee ID 9) as the starting point from the HumanResources.Employee table
- It joins the HumanResources.Employee table with the previous result based on the EmployeeManagerId column to find the manager of each employee. This process continues recursively until no more managers are found
- The last SELECT statement retrieves all columns from the RecursiveHierarchy, which now contains the entire management chain leading to Patricia Doyle

**Question 5:**

**\-- 5-1**

\-- Create a view that returns the total qty

\-- for each employee and year

\-- Tables involved: Sales.Orders and Sales.OrderDetails

  

\-- Desired output when running:

\-- SELECT \* FROM  Sales.VEmpOrders ORDER BY empid, orderyear

empid       orderyear   qty

\----------- ----------- -----------

1           2020        1620

1           2021        3877

1           2022        2315

2           2020        1085

2           2021        2604

2           2022        2366

3           2020        940

3           2021        4436

3           2022        2476

4           2020        2212

4           2021        5273

4           2022        2313

5           2020        778

5           2021        1471

5           2022        787

6           2020        963

6           2021        1738

6           2022        826

7           2020        485

7           2021        2292

7           2022        1877

8           2020        923

8           2021        2843

8           2022        2147

9           2020        575

9           2021        955

9           2022        1140

In [119]:
--Query using TSQLV6 database
USE TSQLV6;
GO

CREATE VIEW Sales.VEmpOrders AS
SELECT O.empid,
    YEAR(O.OrderDate) AS orderyear,
    SUM(OD.qty) AS qty
FROM 
    Sales.Orders AS O
JOIN 
    Sales.OrderDetails OD ON O.OrderID = OD.OrderID
GROUP BY 
    O.empid, YEAR(O.OrderDate);

Commands completed successfully.

Commands completed successfully.

Total execution time: 00:00:00.075

In [120]:
--Query using TSQLV6 database
USE TSQLV6;

SELECT * FROM  Sales.VEmpOrders ORDER BY empid, orderyear

(27 rows affected)

Total execution time: 00:00:00.153

empid orderyear qty 1 2020 1620 1 2021 3877 1 2022 2315 2 2020 1085 2 2021 2604 2 2022 2366 3 2020 940 3 2021 4436 3 2022 2476 4 2020 2212 4 2021 5273 4 2022 2313 5 2020 778 5 2021 1471 5 2022 787 6 2020 963 6 2021 1738 6 2022 826 7 2020 485 7 2021 2292 7 2022 1877 8 2020 923 8 2021 2843 8 2022 2147 9 2020 575 9 2021 955 9 2022 1140

### **Explanation for the TSQLV6 Query**

**Proposition:** Create a view named Sales.VEmpOrder providing a summary of employee orders

**Tables:**

- Sales.Orders aliased as O
- Sales.OrderDetails aliased as OD

**Columns:**

- O.empid
- orderyear
- qty

**Predicate:**

-  Joins Sales.Orders and Sales.OrderDetails on their respective OrderID columns (O.OrderID = OD.OrderID)
- It groups the results by O.empid and YEAR(O.OrderDate) to aggregate the quantity of orders for each employee and year
- The aggregated result is stored in the view Sales.VEmpOrder

In [17]:
-- Query using Northwinds2022TSQLV7 database 
USE Northwinds2022TSQLV7;
GO

CREATE VIEW Sales.VEmpOrders AS
SELECT O.EmployeeId,
    YEAR(O.OrderDate) AS orderyear,
    SUM(OD.Quantity) AS qty
FROM 
    Sales.[Order] AS O
JOIN 
    Sales.[OrderDetail] OD ON O.OrderID = OD.OrderID
GROUP BY 
    O.EmployeeId, YEAR(O.OrderDate);

Commands completed successfully.

Commands completed successfully.

Total execution time: 00:00:00.051

In [35]:
-- Query using Northwinds2022TSQLV7 database 
USE Northwinds2022TSQLV7;

SELECT * FROM  Sales.[VEmpOrders] ORDER BY employeeid, orderyear

(27 rows affected)

Total execution time: 00:00:00.030

EmployeeId orderyear qty 1 2014 1620 1 2015 3877 1 2016 2315 2 2014 1085 2 2015 2604 2 2016 2366 3 2014 940 3 2015 4436 3 2016 2476 4 2014 2212 4 2015 5273 4 2016 2313 5 2014 778 5 2015 1471 5 2016 787 6 2014 963 6 2015 1738 6 2016 826 7 2014 485 7 2015 2292 7 2016 1877 8 2014 923 8 2015 2843 8 2016 2147 9 2014 575 9 2015 955 9 2016 1140

### **Explanation for the Northwinds2022TSQLV7 Query**

**Proposition:** Create a view Sales.VEmpOrders that provides a summary of the employee orders with EmployeeId, orderyear, and Quantity

**Tables:**

- Sales.\[Order\] aliased as O
- Sales.\[OrderDetail\] aliased as OD

**Columns:**

- O.EmployeeId
- Year{O.OrderDate)
- SUM(OD.Quantity)
- Sales.\[OrderDetail\]

**Predicate:**

-  Joins Sales.\[Order\] and Sales.\[OrderDetail\] on their respective OrderID columns (O.OrderID = OD.OrderID)
- Groups the results by O.EmployeeId and YEAR(O.OrderDate) to aggregate the quantity of orders for each employee and year
- The result is stored in the view Sales.VEmpOrders.

**\-- 5-2**

\-- Write a query against Sales.VEmpOrders

\-- that returns the running qty for each employee and year

\-- Tables involved: TSQLV6 database, Sales.VEmpOrders view

  

\-- Desired output:

empid       orderyear   qty         runqty

\----------- ----------- ----------- -----------

1           2020        1620        1620

1           2021        3877        5497

1           2022        2315        7812

2           2020        1085        1085

2           2021        2604        3689

2           2022        2366        6055

3           2020        940         940

3           2021        4436        5376

3           2022        2476        7852

4           2020        2212        2212

4           2021        5273        7485

4           2022        2313        9798

5           2020        778         778

5           2021        1471        2249

5           2022        787         3036

6           2020        963         963

6           2021        1738        2701

6           2022        826         3527

7           2020        485         485

7           2021        2292        2777

7           2022        1877        4654

8           2020        923         923

8           2021        2843        3766

8           2022        2147        5913

9           2020        575         575

9           2021        955         1530

9           2022        1140        2670

In [121]:
--Query using TSQLV6 database
USE TSQLV6;

SELECT 
    empid,
    orderyear,
    qty,
    SUM(qty) OVER (PARTITION BY empid ORDER BY orderyear) AS runqty
FROM 
    Sales.VEmpOrders
ORDER BY 
    empid, orderyear;

(27 rows affected)

Total execution time: 00:00:00.035

empid orderyear qty runqty 1 2020 1620 1620 1 2021 3877 5497 1 2022 2315 7812 2 2020 1085 1085 2 2021 2604 3689 2 2022 2366 6055 3 2020 940 940 3 2021 4436 5376 3 2022 2476 7852 4 2020 2212 2212 4 2021 5273 7485 4 2022 2313 9798 5 2020 778 778 5 2021 1471 2249 5 2022 787 3036 6 2020 963 963 6 2021 1738 2701 6 2022 826 3527 7 2020 485 485 7 2021 2292 2777 7 2022 1877 4654 8 2020 923 923 8 2021 2843 3766 8 2022 2147 5913 9 2020 575 575 9 2021 955 1530 9 2022 1140 2670

**Explanation for the TSQLV6 Query**

**Proposition:** Retrieve data from the Sales.VEmpOrders view and calculate the runqty for each employee and the year.

**Table:**

- Sales.VEmpOrders

**Columns:**

- empid
- orderyear
- qty
- runqty

**Predicate:**

- SUM(qty) OVER (PARTITION BY empid ORDER BY orderyear)

<span style="color: var(--vscode-foreground);">This window function calculates the running sum of the qty column partitioned by empid and ordered by orderyear. It calculates the sum of qty for each row, considering only rows with the same empid and rows ordered by orderyear.</span>

In [20]:
-- Query using Northwinds2022TSQLV7 database 
USE Northwinds2022TSQLV7;

SELECT 
    EmployeeId,
    orderyear,
    qty,
    SUM(qty) OVER (PARTITION BY employeeid ORDER BY orderyear ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS runqty
FROM 
    Sales.[VEmpOrders]
ORDER BY 
    EmployeeId, orderyear;

(27 rows affected)

Total execution time: 00:00:00.019

EmployeeId orderyear qty runqty 1 2014 1620 1620 1 2015 3877 5497 1 2016 2315 7812 2 2014 1085 1085 2 2015 2604 3689 2 2016 2366 6055 3 2014 940 940 3 2015 4436 5376 3 2016 2476 7852 4 2014 2212 2212 4 2015 5273 7485 4 2016 2313 9798 5 2014 778 778 5 2015 1471 2249 5 2016 787 3036 6 2014 963 963 6 2015 1738 2701 6 2016 826 3527 7 2014 485 485 7 2015 2292 2777 7 2016 1877 4654 8 2014 923 923 8 2015 2843 3766 8 2016 2147 5913 9 2014 575 575 9 2015 955 1530 9 2016 1140 2670

### **Explanation for the Northwinds2022TSQLV7 Query**

**Proposition:** <span style="color: var(--vscode-foreground);">Retrieve data from the Sales.VEmpOrders view and calculate the runqty for each employee and the year.</span>

**Tables:**

- Sales.\[VEmpOrders\]

**Columns:**

- EmployeeId
- orderyear
- qty
- runqty

**Predicate:**

- This query does not directly involve tables or columns. It operates on data already aggregated and provided by the Sales.\[VEmpOrders\] view. It calculates the running total (runqty) without filtering or conditioning the data further.

**Question 6:**

**\-- 6-1**

\-- Create an inline function that accepts as inputs

\-- a supplier id (@supid AS INT), 

\-- and a requested number of products (@n AS INT)

\-- The function should return @n products with the highest unit prices

\-- that are supplied by the given supplier id

\-- Tables involved: Production.Products

  

\-- Desired output when issuing the following query:

\-- SELECT \* FROM Production.TopProducts(5, 2)

  

productid   productname                              unitprice

\----------- ---------------------------------------- ---------------------

12          Product OSFNS                            38.00

11          Product QMVUN                            21.00

In [125]:
--Query using TSQLV6 database
USE TSQLV6;

CREATE FUNCTION Production.TopProducts 
(
    @supid INT,
    @n INT
)
RETURNS TABLE
AS
RETURN
(
    SELECT TOP (@n) productid, productname, unitprice
    FROM Production.Products
    WHERE supplierid = @supid
    ORDER BY unitprice DESC
);


Commands completed successfully.

Total execution time: 00:00:00.016

In [128]:
--Query using TSQLV6 database
USE TSQLV6;

SELECT * FROM Production.TopProducts(5, 2)

(2 rows affected)

Total execution time: 00:00:00.041

productid productname unitprice 12 Product OSFNS 38.00 11 Product QMVUN 21.00

**Explanation for the TSQLV6 Query**

**Proposition:** Create a function that accepts two input parameters @supid (supplier ID) and @n (requested number of products) and returns a table with the top @n products having the highest unit prices supplied by the given supplier ID

**Table:**

- Production.Products

**Columns:**

- productid
- productname
- unitprice
- supplierid

**Predicate:**

- The function filters the rows from Production.Products table based on the supplierid column matching the input parameter @supid
- It selects the top @n products with the highest unit prices and orders them in descending order by unitprice

In [48]:
-- Query using Northwinds2022TSQLV7 database 
USE Northwinds2022TSQLV7;

CREATE FUNCTION Production.TopProducts
(
    @supid INT,
    @n INT
)
RETURNS TABLE
AS
RETURN
(
    SELECT TOP (@n) ProductId, ProductName, UnitPrice
    FROM Production.[Products]
    WHERE SupplierId = @supid
    ORDER BY UnitPrice DESC
);

Commands completed successfully.

Total execution time: 00:00:00.028

In [54]:
-- Query using Northwinds2022TSQLV7 database 
USE Northwinds2022TSQLV7;

SELECT * FROM Production.[TopProducts](5, 2)

(2 rows affected)

Total execution time: 00:00:00.030

productid productname unitprice 12 Product OSFNS 38.00 11 Product QMVUN 21.00

### **Explanation for the Northwinds2022TSQLV7 Query**

**Proposition:** The function accepts two inputs @supid (supplier ID) and @n (requested number of products) and returns a table containing the top @n products with the highest unit prices supplied by the given supplier ID

**Tables:**

- Production.\[Products\]

**Columns:**

- ProductID
- ProductName
- UnitPrice

**Predicate:**

- Filters rows from the Production.\[Products\] table based on the SupplierId column matching the input parameter @supid
- It selects the top @n products with the highest unit prices from the filtered results and orders them in descending order by UnitPrice
- Function then returns this filtered and ordered result set as a table.

**\-- 6-2**

\-- Using the CROSS APPLY operator

\-- and the function you created in exercise 6-1,

\-- return, for each supplier, the two most expensive products

  

\-- Desired output 

supplierid  companyname     productid   productname     unitprice

\----------- --------------- ----------- --------------- ----------

8           Supplier BWGYE  20          Product QHFFP   81.00

8           Supplier BWGYE  68          Product TBTBL   12.50

20          Supplier CIYNM  43          Product ZZZHR   46.00

20          Supplier CIYNM  44          Product VJIEO   19.45

23          Supplier ELCRN  49          Product FPYPN   20.00

23          Supplier ELCRN  76          Product JYGFE   18.00

5           Supplier EQPNC  12          Product OSFNS   38.00

5           Supplier EQPNC  11          Product QMVUN   21.00

In [131]:
--Query using TSQLV6 database
USE TSQLV6;

SELECT 
    s.supplierid,
    s.companyname,
    tp.productid,
    tp.productname,
    tp.unitprice
FROM 
    Production.Suppliers s
CROSS APPLY
    Production.TopProducts(s.supplierid, 2) tp;


(55 rows affected)

Total execution time: 00:00:00.024

supplierid companyname productid productname unitprice 8 Supplier BWGYE 20 Product QHFFP 81.00 8 Supplier BWGYE 68 Product TBTBL 12.50 20 Supplier CIYNM 43 Product ZZZHR 46.00 20 Supplier CIYNM 44 Product VJIEO 19.45 23 Supplier ELCRN 49 Product FPYPN 20.00 23 Supplier ELCRN 76 Product JYGFE 18.00 5 Supplier EQPNC 12 Product OSFNS 38.00 5 Supplier EQPNC 11 Product QMVUN 21.00 25 Supplier ERVYZ 55 Product YYWRT 24.00 25 Supplier ERVYZ 54 Product QAQRL 7.45 22 Supplier FNUXM 48 Product MYNXN 12.75 22 Supplier FNUXM 47 Product EZZPR 9.50 7 Supplier GQRCV 18 Product CKEDC 62.50 7 Supplier GQRCV 63 Product ICKNK 43.90 19 Supplier JDNUG 40 Product YZIXQ 18.40 19 Supplier JDNUG 41 Product TTEEX 9.65 24 Supplier JNNES 51 Product APITJ 53.00 24 Supplier JNNES 53 Product BKGEA 32.80 14 Supplier KEREV 72 Product GEEOO 34.80 14 Supplier KEREV 32 Product NUNAW 32.00 18 Supplier LVJUA 38 Product QDOMO 263.50 18 Supplier LVJUA 39 Product LSOFL 18.00 15 Supplier NZLIF 69 Product COAXA 36.00 15 Supplier NZLIF 71 Product MYMOI 21.50 28 Supplier OAVQT 59 Product UKXRI 55.00 28 Supplier OAVQT 60 Product WHBYK 34.00 29 Supplier OGLRK 62 Product WUXYK 49.30 29 Supplier OGLRK 61 Product XYZPE 28.50 4 Supplier QOVFD 9 Product AOZBW 97.00 4 Supplier QOVFD 10 Product YHXGE 31.00 9 Supplier QQYEU 22 Product CPHFY 21.00 9 Supplier QQYEU 23 Product JLUDZ 9.00 6 Supplier QWUSF 14 Product PWCJB 23.25 6 Supplier QWUSF 15 Product KSZOI 15.50 17 Supplier QZGUF 37 Product EVFFA 26.00 17 Supplier QZGUF 36 Product GMKIJ 19.00 3 Supplier STUAZ 8 Product WVJFP 40.00 3 Supplier STUAZ 7 Product HMLNI 30.00 12 Supplier SVIYA 29 Product VJXYN 123.79 12 Supplier SVIYA 28 Product OFBNT 45.60 1 Supplier SWRXU 2 Product RECZE 19.00 1 Supplier SWRXU 1 Product HHYDP 18.00 13 Supplier TEGSC 30 Product LYERX 25.89 16 Supplier UHZRG 35 Product NEVTJ 18.00 16 Supplier UHZRG 67 Product XLXQF 14.00 10 Supplier UNAHG 24 Product QOGNU 4.50 2 Supplier VHQZD 4 Product KSBRM 22.00 2 Supplier VHQZD 5 Product EPEIM 21.35 21 Supplier XOXZA 46 Product CBRRL 12.00 21 Supplier XOXZA 45 Product AQOKR 9.50 11 Supplier ZPYVS 27 Product SMIOH 43.90 11 Supplier ZPYVS 26 Product HLGZA 31.23 27 Supplier ZRYDZ 58 Product ACRVI 13.25 26 Supplier ZWZDM 56 Product VKCMF 38.00 26 Supplier ZWZDM 57 Product OVLQI 19.50

**Explanation for the TSQLV6 Query**

**Proposition:** Retrieves the two most expensive products for each supplier from Production.Suppliers table using a user-defined table-valued function named Production.TopProducts and is combined using the CROSS APPLY operator

**Table:**

- Production.Suppliers

**Columns:**

- supplierid
- companyname
- productid
- productname
- unitprice

**Predicate:**

- The function returns the two most expensive products for each supplier
- The result has columns from both Suppliers table and the result of the function call, providing details about each supplier and their top two products sorted by unit price

In [58]:
-- Query using Northwinds2022TSQLV7 database 
USE Northwinds2022TSQLV7;

SELECT 
    s.SupplierId,
    s.CompanyName,
    tp.ProductId,
    tp.ProductName,
    tp.UnitPrice
FROM 
    Production.[Supplier] s
CROSS APPLY
    Production.[TopProducts](s.SupplierId, 2) tp;



(55 rows affected)

Total execution time: 00:00:00.036

SupplierId CompanyName ProductId ProductName UnitPrice 8 Supplier BWGYE 20 Product QHFFP 81.00 8 Supplier BWGYE 68 Product TBTBL 12.50 20 Supplier CIYNM 43 Product ZZZHR 46.00 20 Supplier CIYNM 44 Product VJIEO 19.45 23 Supplier ELCRN 49 Product FPYPN 20.00 23 Supplier ELCRN 76 Product JYGFE 18.00 5 Supplier EQPNC 12 Product OSFNS 38.00 5 Supplier EQPNC 11 Product QMVUN 21.00 25 Supplier ERVYZ 55 Product YYWRT 24.00 25 Supplier ERVYZ 54 Product QAQRL 7.45 22 Supplier FNUXM 48 Product MYNXN 12.75 22 Supplier FNUXM 47 Product EZZPR 9.50 7 Supplier GQRCV 18 Product CKEDC 62.50 7 Supplier GQRCV 63 Product ICKNK 43.90 19 Supplier JDNUG 40 Product YZIXQ 18.40 19 Supplier JDNUG 41 Product TTEEX 9.65 24 Supplier JNNES 51 Product APITJ 53.00 24 Supplier JNNES 53 Product BKGEA 32.80 14 Supplier KEREV 72 Product GEEOO 34.80 14 Supplier KEREV 32 Product NUNAW 32.00 18 Supplier LVJUA 38 Product QDOMO 263.50 18 Supplier LVJUA 39 Product LSOFL 18.00 15 Supplier NZLIF 69 Product COAXA 36.00 15 Supplier NZLIF 71 Product MYMOI 21.50 28 Supplier OAVQT 59 Product UKXRI 55.00 28 Supplier OAVQT 60 Product WHBYK 34.00 29 Supplier OGLRK 62 Product WUXYK 49.30 29 Supplier OGLRK 61 Product XYZPE 28.50 4 Supplier QOVFD 9 Product AOZBW 97.00 4 Supplier QOVFD 10 Product YHXGE 31.00 9 Supplier QQYEU 22 Product CPHFY 21.00 9 Supplier QQYEU 23 Product JLUDZ 9.00 6 Supplier QWUSF 14 Product PWCJB 23.25 6 Supplier QWUSF 15 Product KSZOI 15.50 17 Supplier QZGUF 37 Product EVFFA 26.00 17 Supplier QZGUF 36 Product GMKIJ 19.00 3 Supplier STUAZ 8 Product WVJFP 40.00 3 Supplier STUAZ 7 Product HMLNI 30.00 12 Supplier SVIYA 29 Product VJXYN 123.79 12 Supplier SVIYA 28 Product OFBNT 45.60 1 Supplier SWRXU 2 Product RECZE 19.00 1 Supplier SWRXU 1 Product HHYDP 18.00 13 Supplier TEGSC 30 Product LYERX 25.89 16 Supplier UHZRG 35 Product NEVTJ 18.00 16 Supplier UHZRG 67 Product XLXQF 14.00 10 Supplier UNAHG 24 Product QOGNU 4.50 2 Supplier VHQZD 4 Product KSBRM 22.00 2 Supplier VHQZD 5 Product EPEIM 21.35 21 Supplier XOXZA 46 Product CBRRL 12.00 21 Supplier XOXZA 45 Product AQOKR 9.50 11 Supplier ZPYVS 27 Product SMIOH 43.90 11 Supplier ZPYVS 26 Product HLGZA 31.23 27 Supplier ZRYDZ 58 Product ACRVI 13.25 26 Supplier ZWZDM 56 Product VKCMF 38.00 26 Supplier ZWZDM 57 Product OVLQI 19.50

### **Explanation for the Northwinds2022TSQLV7 Query**

**Proposition:** Retrieve data that contains the two most expensive products for each supplier and combine the two using the CROSS APPLY operator.

**Tables:**

- Production.Supplier

**Columns:**

- <span style="color: var(--vscode-foreground);">SupplierId</span>

- CompanyName
- ProductId
- ProductName
- UnitPrice

**Predicate:**

- The function returns the two most expensive productsfor each supplier
- Results in a dataset containing columns from both the Supplier table and the result of the function call

## **Part 2:**

**Chapter 4 | Lines 219-276**

\---------------------------------------------------------------------

\-- Misbehaving Subqueries

\---------------------------------------------------------------------

\---------------------------------------------------------------------

\-- NULL Trouble

\---------------------------------------------------------------------

**Preposition:**

Write a query that identifies customers from the  Sales.Customers table who haven't placed any orders and if the custid is NULL delete it from the Sales.Orders table.

**Tables:**

- Sales.Orders

**Columns:**

- custid
- companyname

**Predicate:** 

- Selects customer IDs (custid) and company names (companyname) from the Sales.Customers table where there doesn't exist any corresponding record in the Sales.Orders table with the same custid
- The DELETE statement removes any records from the Sales.Orders table where the custid is NULL. This ensures data integrity by removing invalid or incomplete records.

In [21]:
--Query using TSQLV6 database
USE TSQLV6;

-- Using NOT EXISTS
SELECT custid, companyname
FROM Sales.Customers AS C
WHERE NOT EXISTS
  (SELECT * 
   FROM Sales.Orders AS O
   WHERE O.custid = C.custid);

-- Cleanup
DELETE FROM Sales.Orders WHERE custid IS NULL;


--Query using Northwinds2022TSQLV7
USE Northwinds2022TSQLV7;

-- Using NOT EXISTS
SELECT CustomerId, CustomerCompanyName
FROM Sales.[Customer] AS C
WHERE NOT EXISTS
  (SELECT * 
   FROM Sales.[Order] AS O
   WHERE O.CustomerId = C.CustomerId);

-- Cleanup
DELETE FROM Sales.[Order] WHERE CustomerId IS NULL;

(2 rows affected)

(0 rows affected)

(2 rows affected)

(0 rows affected)

Total execution time: 00:00:00.022

custid companyname 22 Customer DTDMN 57 Customer WVAXS

CustomerId CustomerCompanyName 22 Customer DTDMN 57 Customer WVAXS

\---------------------------------------------------------------------

\-- Substitution Error in a Subquery Column Name

\---------------------------------------------------------------------

**Preposition:** 

Create a table Sales.MyShippers to store information about shippers and populate it with data for three shippers, including their shipper IDs, company names, and phone numbers.

**Tables:**

- Sales.MyShippers

**Columns:**

- shipper\_id
- companyname
- phone

**Predicate:** 

- The table inserts  three records into the table to populate it with information about different shippers and each record includes shipper ID, company name, and phone number for a shipper

In [24]:
--Query using TSQLV6 database
USE TSQLV6;
-- Create and populate table Sales.MyShippers

DROP TABLE IF EXISTS Sales.MyShippers;

CREATE TABLE Sales.MyShippers

(

  shipper_id  INT          NOT NULL,

  companyname NVARCHAR(40) NOT NULL,

  phone       NVARCHAR(24) NOT NULL,

  CONSTRAINT PK_MyShippers PRIMARY KEY(shipper_id)

);

INSERT INTO Sales.MyShippers(shipper_id, companyname, phone)

  VALUES(1, N'Shipper GVSUA', N'(503) 555-0137'),

      (2, N'Shipper ETYNR', N'(425) 555-0136'),

(3, N'Shipper ZHISN', N'(415) 555-0138');

(3 rows affected)

Total execution time: 00:00:00.052

In [26]:
--Query using TSQLV6 database
USE TSQLV6;

SELECT * FROM Sales.MyShippers;

(3 rows affected)

Total execution time: 00:00:00.013

shipper_id companyname phone 1 Shipper GVSUA (503) 555-0137 2 Shipper ETYNR (425) 555-0136 3 Shipper ZHISN (415) 555-0138

**Preposition:** 

Retreive information about shippers who shipped orders to customer 43 from the Sales.Shippers table.

**Tables:**

- <span style="color: var(--vscode-foreground);">Sales.Shippers</span>
    
- Sales.Orders
    

**Columns:**

- shipperid
- companyname
- custid

**Predicate:** 

- It tales the shipper IDs and company names where it exists and the subquery retrieves the shipper IDs from the Sales.Orders table where the custid  is 43.

In [27]:
--Query using TSQLV6 database
USE TSQLV6;
-- Shippers who shipped orders to customer 43
-- Bug
SELECT shipperid, companyname
FROM Sales.Shippers
WHERE shipperid IN
  (SELECT shipperid
   FROM Sales.Orders
   WHERE custid = 43);

--Query using Northwinds2022TSQLV7
USE Northwinds2022TSQLV7;
-- Shippers who shipped orders to customer 43
-- Bug
SELECT ShipperId, ShipperCompanyName
FROM Sales.[Shipper]
WHERE Shipperid IN
  (SELECT Shipperid
   FROM Sales.[Order]
   WHERE CustomerId = 43);

-----------------------------------------------------------------

--Query using TSQLV6 database
USE TSQLV6;
-- The safe way using aliases, bug identified
SELECT shipperid, companyname
FROM Sales.Shippers
WHERE shipperid IN
  (SELECT O.shipperid
   FROM Sales.Orders AS O
   WHERE O.custid = 43);

--Query using Northwinds2022TSQLV7
USE Northwinds2022TSQLV7;
-- The safe way using aliases, bug identified
SELECT Shipperid, ShipperCompanyName
FROM Sales.[Shipper]
WHERE Shipperid IN
  (SELECT O.Shipperid
   FROM Sales.[Order] AS O
   WHERE O.CustomerId = 43);

---------------------------------------------------------------

--Query using TSQLV6 database
USE TSQLV6;
-- Bug corrected
SELECT shipperid, companyname
FROM Sales.Shippers
WHERE shipperid IN
  (SELECT O.shipperid
   FROM Sales.Orders AS O
   WHERE O.custid = 43);

--Query using Northwinds2022TSQLV7
USE Northwinds2022TSQLV7;
-- Bug corrected
SELECT Shipperid, ShipperCompanyName
FROM Sales.[Shipper]
WHERE Shipperid IN
  (SELECT O.Shipperid
   FROM Sales.[Order] AS O
   WHERE O.CustomerId = 43);

-- Cleanup
DROP TABLE IF EXISTS Sales.MyShippers;

(2 rows affected)

(2 rows affected)

(2 rows affected)

(2 rows affected)

(2 rows affected)

(2 rows affected)

Total execution time: 00:00:00.056

shipperid companyname 2 Shipper ETYNR 3 Shipper ZHISN

ShipperId ShipperCompanyName 2 Shipper ETYNR 3 Shipper ZHISN

shipperid companyname 2 Shipper ETYNR 3 Shipper ZHISN

Shipperid ShipperCompanyName 2 Shipper ETYNR 3 Shipper ZHISN

shipperid companyname 2 Shipper ETYNR 3 Shipper ZHISN

Shipperid ShipperCompanyName 2 Shipper ETYNR 3 Shipper ZHISN

**Chapter 5 | Lines 383-490**

\---------------------------------------------------------------------

\-- CHECK OPTION

\---------------------------------------------------------------------

**Preposition:** 

Insert a new row into Sales.Customers table through a view. Then, cleanup by deleting rows from the Sales.Customers table where the custid is greater than 91 and dropping the view Sales.USACusts.

**Tables:**

- Sales.Customers

**Columns:**

- companyname
- contactname
- contacttitle
- address
- city
- region
- postalcode
- country
- phone
- fax

**Predicate:**

- Inserts a new row into the Sales.Customers table specifying the vairous columns. Then it will try to delete the rows from the Sales.Customers table where the custid is greater than 91
- It drops the view Sales.USACusts if it exists

In [34]:
--Query using TSQLV6 database
USE TSQLV6;
-- Notice that you can't insert a row through the view
INSERT INTO Sales.Customers(
  companyname, contactname, contacttitle, address,
  city, region, postalcode, country, phone, fax)
 VALUES(
  N'Customer FGHIJ', N'Contact FGHIJ', N'Title FGHIJ', N'Address FGHIJ',
  N'London', NULL, N'12345', N'UK', N'012-3456789', N'012-3456789');

GO

-- Cleanup
DELETE FROM Sales.Customers
WHERE custid > 91;

DROP VIEW IF EXISTS Sales.USACusts;
GO


--Query using Northwinds2022TSQLV7
USE Northwinds2022TSQLV7;
-- Notice that you can't insert a row through the view
INSERT INTO Sales.[Customer](
  CustomerCompanyName, CustomerContactName, CustomerContactTitle, CustomerAddress,
  CustomerCity, CustomerRegion, CustomerPostalCode, CustomerCountry, CustomerPhoneNumber, CustomerFaxNumber)
 VALUES(
  N'Customer FGHIJ', N'Contact FGHIJ', N'Title FGHIJ', N'Address FGHIJ',
  N'London', NULL, N'12345', N'UK', N'012-3456789', N'012-3456789');

GO

-- Cleanup
DELETE FROM Sales.[Customer]
WHERE customerid > 91;

DROP VIEW IF EXISTS Sales.USACusts;
GO

(1 row affected)

(1 row affected)

(1 row affected)

(1 row affected)

Total execution time: 00:00:00.168

\---------------------------------------------------------------------

\-- Inline User Defined Functions

\---------------------------------------------------------------------

**Preposition:** 

Create or alter a function named dbo.GetCustOrders in the TSQLV6 database to retrieve orders for a specific customer ID. Then test the function by selecting order information for customer ID 1 and joining it with order details.

**Tables:**

- Sales.Orders
- Sales.OrderDetails

**Columns:**

- <span style="color: var(--vscode-foreground);">orderid</span>
- <span style="color: var(--vscode-foreground);">custid</span>
- <span style="color: var(--vscode-foreground);">empid</span>
- <span style="color: var(--vscode-foreground);">orderdate</span>
- <span style="color: var(--vscode-foreground);">requireddate</span>
- shippeddate
- shipperid
- freight
- shipname
- shipaddress
- shipcity
- shipregion
- shippostalcode
- shipcountry

**Predicate:**

- The function named dbo.GetCustOrders returns the order information for a specific customer ID (@cid)
- Then tests the function by ordering the information for customer  ID 1 and joining it with order details to retrieve product information
- Then it cleans up the function by dropping the function dbo.GetCustOrders if it exists

In [35]:
--Query using TSQLV6 database
USE TSQLV6;
GO
-- Creating GetCustOrders function

CREATE OR ALTER FUNCTION dbo.GetCustOrders
  (@cid AS INT) RETURNS TABLE
AS
RETURN
  SELECT orderid, custid, empid, orderdate, requireddate,
    shippeddate, shipperid, freight, shipname, shipaddress, shipcity,
    shipregion, shippostalcode, shipcountry
  FROM Sales.Orders
  WHERE custid = @cid;
GO
-- Test Function
SELECT orderid, custid
FROM dbo.GetCustOrders(1) AS O;

SELECT O.orderid, O.custid, OD.productid, OD.qty
FROM dbo.GetCustOrders(1) AS O
  INNER JOIN Sales.OrderDetails AS OD
    ON O.orderid = OD.orderid;
GO

-- Cleanup
DROP FUNCTION IF EXISTS dbo.GetCustOrders;
GO


--Query using Northwinds2022TSQLV7
USE Northwinds2022TSQLV7;
GO
-- Creating GetCustOrders function

CREATE OR ALTER FUNCTION dbo.GetCustOrders
  (@cid AS INT) RETURNS TABLE
AS
RETURN
  SELECT OrderId, CustomerId, EmployeeId, OrderDate, RequiredDate,
    ShipToDate, shipperid, freight, ShipToName, ShipToAddress, ShipToCity,
    ShipToRegion, ShipToPostalCode, ShipToCountry
  FROM Sales.[Order]
  WHERE CustomerId = @cid;
GO
-- Test Function
SELECT OrderId, CustomerId
FROM dbo.GetCustOrders(1) AS O;

SELECT O.OrderId, O.CustomerId, OD.ProductId, OD.Quantity
FROM dbo.GetCustOrders(1) AS O
  INNER JOIN Sales.[OrderDetail] AS OD
    ON O.OrderId = OD.OrderId;
GO

-- Cleanup
DROP FUNCTION IF EXISTS dbo.GetCustOrders;
GO

Commands completed successfully.

Commands completed successfully.

(6 rows affected)

(12 rows affected)

Commands completed successfully.

Commands completed successfully.

Commands completed successfully.

(6 rows affected)

(12 rows affected)

Commands completed successfully.

Total execution time: 00:00:00.164

orderid custid 10643 1 10692 1 10702 1 10835 1 10952 1 11011 1

orderid custid productid qty 10643 1 28 15 10643 1 39 21 10643 1 46 2 10692 1 63 20 10702 1 3 6 10702 1 76 15 10835 1 59 15 10835 1 77 2 10952 1 6 16 10952 1 28 2 11011 1 58 40 11011 1 71 20

OrderId CustomerId 10643 1 10692 1 10702 1 10835 1 10952 1 11011 1

OrderId CustomerId ProductId Quantity 10643 1 28 15 10643 1 39 21 10643 1 46 2 10692 1 63 20 10702 1 3 6 10702 1 76 15 10835 1 59 15 10835 1 77 2 10952 1 6 16 10952 1 28 2 11011 1 58 40 11011 1 71 20

\---------------------------------------------------------------------

\-- APPLY

\---------------------------------------------------------------------

**Preposition:** 

Retrieves the shipper IDs and employee IDs by performing a cross join and a cross apply operation between the Sales.Shippers table and the HR.Employees table

**Tables:**

- Sales.Shippers
- HR.Employees

**Columns:**

- S.shipperid
- E.empid

**Predicate:**

- A cross join between Sales.Shippers table and the HR.Employees table is performed. It generates all  combinations of shipper IDs and employee IDs
- Each row from Sales.Shippers table is combined with HR.Employees table, resulting in a Cartesian product
- Then a cross apply operation is performed and it filters based on each row of the outer table. 
- Each row from the Sales.Shippers table is combined with only the rows from the HR.Employees table that satisfy certain conditions
- Cross apply results in a constrained result compared to cross join

In [90]:
--Query using TSQLV6 database
USE TSQLV6;

SELECT S.shipperid, E.empid
FROM Sales.Shippers AS S
  CROSS JOIN HR.Employees AS E;

SELECT S.shipperid, E.empid
FROM Sales.Shippers AS S
  CROSS APPLY HR.Employees AS E;


--Query using Northwinds2022TSQLV7
USE Northwinds2022TSQLV7;

SELECT S.ShipperId, E.EmployeeId
FROM Sales.[Shipper] AS S
  CROSS JOIN HumanResources.[Employee] AS E;

SELECT S.Shipperid, E.EmployeeId
FROM Sales.[Shipper] AS S
  CROSS APPLY HumanResources.[Employee] AS E;

(27 rows affected)

(27 rows affected)

(27 rows affected)

(27 rows affected)

Total execution time: 00:00:00.045

shipperid empid 1 1 1 2 1 3 1 4 1 5 1 6 1 7 1 8 1 9 2 1 2 2 2 3 2 4 2 5 2 6 2 7 2 8 2 9 3 1 3 2 3 3 3 4 3 5 3 6 3 7 3 8 3 9

shipperid empid 1 1 1 2 1 3 1 4 1 5 1 6 1 7 1 8 1 9 2 1 2 2 2 3 2 4 2 5 2 6 2 7 2 8 2 9 3 1 3 2 3 3 3 4 3 5 3 6 3 7 3 8 3 9

ShipperId EmployeeId 1 1 1 2 1 3 1 4 1 5 1 6 1 7 1 8 1 9 2 1 2 2 2 3 2 4 2 5 2 6 2 7 2 8 2 9 3 1 3 2 3 3 3 4 3 5 3 6 3 7 3 8 3 9

ShipperId EmployeeId 1 1 1 2 1 3 1 4 1 5 1 6 1 7 1 8 1 9 2 1 2 2 2 3 2 4 2 5 2 6 2 7 2 8 2 9 3 1 3 2 3 3 3 4 3 5 3 6 3 7 3 8 3 9

**Preposition:** 

Retrieve the three most recent orders for each customer from the Sales.Customers table by performing a cross apply operation with a subquery selecting the top three orders for each customer from the Sales.Orders table

**Tables:**

- Sales.Customers
- Sales.Orders

**Columns:**

- <span style="color: var(--vscode-foreground);">custid</span>
    
- orderid
    
- orderdate
    
- empid
    
- requireddate
    

**Predicate:**

- A cross apply operation is used to combine each row from the Sales.Customers table with the top three most recent orders from the corresponding customer from the Sales.Orders table
- The subquery inside takes the top three orders for each customer and is filtered based on the customer ID (O.custid = C.custid) and orders the reults by the order date in descending order (latest orders first), and then by order ID in descending order (to break ties if multiple orders have the same order date)

In [92]:
--Query using TSQLV6 database
USE TSQLV6;

-- 3 most recent orders for each customer
SELECT C.custid, A.orderid, A.orderdate
FROM Sales.Customers AS C
  CROSS APPLY
    (SELECT TOP (3) orderid, empid, orderdate, requireddate 
     FROM Sales.Orders AS O
     WHERE O.custid = C.custid
     ORDER BY orderdate DESC, orderid DESC) AS A;

--Query using Northwinds2022TSQLV7
USE Northwinds2022TSQLV7;

-- 3 most recent orders for each customer
SELECT C.Customerid, A.OrderId, A.OrderDate
FROM Sales.[Customer] AS C
  CROSS APPLY
    (SELECT TOP (3) Orderid, e=EmployeeId, OrderDate, RequiredDate 
     FROM Sales.[Order] AS O
     WHERE O.CustomerId = C.CustomerId
     ORDER BY OrderDate DESC, OrderId DESC) AS A;

(263 rows affected)

(263 rows affected)

Total execution time: 00:00:00.094

custid orderid orderdate 1 11011 2022-04-09 1 10952 2022-03-16 1 10835 2022-01-15 2 10926 2022-03-04 2 10759 2021-11-28 2 10625 2021-08-08 3 10856 2022-01-28 3 10682 2021-09-25 3 10677 2021-09-22 4 11016 2022-04-10 4 10953 2022-03-16 4 10920 2022-03-03 5 10924 2022-03-04 5 10875 2022-02-06 5 10866 2022-02-03 6 11058 2022-04-29 6 10956 2022-03-17 6 10853 2022-01-27 7 10826 2022-01-12 7 10679 2021-09-23 7 10628 2021-08-12 8 10970 2022-03-24 8 10801 2021-12-29 8 10326 2020-10-10 9 11076 2022-05-06 9 10940 2022-03-11 9 10932 2022-03-06 11 11023 2022-04-14 11 10947 2022-03-13 11 10943 2022-03-11 12 11054 2022-04-28 12 10937 2022-03-10 12 10881 2022-02-11 13 10259 2020-07-18 14 11041 2022-04-22 14 11029 2022-04-16 14 10966 2022-03-20 16 10848 2022-01-23 16 10462 2021-03-03 16 10435 2021-02-04 17 11067 2022-05-04 17 11036 2022-04-20 17 10825 2022-01-09 18 10890 2022-02-16 18 10683 2021-09-26 18 10609 2021-07-24 19 11056 2022-04-28 19 11047 2022-04-24 19 11024 2022-04-15 20 11072 2022-05-05 20 11017 2022-04-13 20 11008 2022-04-08 23 10789 2021-12-22 23 10763 2021-12-03 23 10634 2021-08-15 24 11050 2022-04-27 24 11001 2022-04-06 24 10993 2022-04-01 25 11012 2022-04-09 25 10929 2022-03-05 25 10859 2022-01-29 26 10971 2022-03-24 26 10860 2022-01-29 26 10671 2021-09-17 27 11060 2022-04-30 27 11026 2022-04-15 27 10807 2021-12-31 28 10963 2022-03-19 28 10664 2021-09-10 28 10604 2021-07-18 29 10928 2022-03-05 29 10887 2022-02-13 29 10568 2021-06-13 30 11037 2022-04-21 30 11009 2022-04-08 30 10948 2022-03-13 39 11028 2022-04-16 39 10893 2022-02-18 39 10849 2022-01-23 40 10973 2022-03-24 40 10972 2022-03-24 40 10927 2022-03-05 41 11051 2022-04-27 41 10923 2022-03-03 41 10832 2022-01-14 44 11070 2022-05-05 44 10934 2022-03-09 44 10891 2022-02-17 49 10950 2022-03-16 49 10939 2022-03-10 49 10818 2022-01-07 50 11004 2022-04-07 50 10978 2022-03-26 50 10896 2022-02-19 52 10945 2022-03-12 52 10779 2021-12-16 52 10699 2021-10-09 53 11057 2022-04-29 53 10752 2021-11-24 53 10517 2021-04-24 54 10986 2022-03-30 54 10958 2022-03-18 54 10898 2022-02-20 56 11020 2022-04-14 56 10999 2022-04-03 56 10833 2022-01-15 58 11073 2022-05-05 58 10995 2022-04-02 58 10502 2021-04-10 59 11053 2022-04-27 59 10844 2022-01-21 59 10747 2021-11-19 60 11007 2022-04-08 60 10477 2021-03-17 60 10433 2021-02-03 63 11021 2022-04-14 63 10996 2022-04-02 63 10991 2022-04-01 64 11019 2022-04-13 64 10916 2022-02-27 64 10828 2022-01-13 66 11062 2022-04-30 66 11010 2022-04-09 66 10942 2022-03-11 68 11075 2022-05-06 68 11033 2022-04-17 68 10951 2022-03-16 69 11013 2022-04-09 69 10917 2022-03-02 69 10306 2020-09-16 70 11015 2022-04-10 70 10909 2022-02-26 70 10831 2022-01-14 72 10869 2022-02-04 72 10804 2021-12-30 72 10800 2021-12-26 73 11074 2022-05-06 73 10802 2021-12-29 73 10669 2021-09-15 74 11043 2022-04-22 74 10964 2022-03-20 74 10907 2022-02-25 76 11038 2022-04-21 76 11035 2022-04-20 76 10930 2022-03-06 79 10967 2022-03-23 79 10608 2021-07-23 79 10548 2021-05-26 80 11069 2022-05-04 80 10915 2022-02-27 80 10842 2022-01-20 83 10994 2022-04-02 83 10946 2022-03-12 83 10921 2022-03-03 84 10850 2022-01-23 84 10843 2022-01-21 84 10814 2022-01-05 85 10739 2021-11-12 85 10737 2021-11-11 85 10295 2020-09-02 86 11046 2022-04-23 86 10668 2021-09-15 86 10651 2021-09-01 87 11025 2022-04-15 87 10781 2021-12-17 87 10750 2021-11-21 90 11005 2022-04-07 90 10910 2022-02-26 90 10879 2022-02-10 91 11044 2022-04-23 91 10998 2022-04-03 91 10906 2022-02-25 55 11034 2022-04-20 55 10965 2022-03-20 55 10855 2022-01-27 10 11048 2022-04-24 10 11045 2022-04-23 10 11027 2022-04-16 42 10810 2022-01-01 42 10620 2021-08-05 42 10495 2021-04-03 45 10884 2022-02-12 45 10735 2021-11-10 45 10719 2021-10-27 37 11063 2022-04-30 37 10985 2022-03-30 37 10912 2022-02-26 33 10785 2021-12-18 33 10268 2020-07-30 71 11064 2022-05-01 71 11031 2022-04-17 71 11030 2022-04-17 38 10933 2022-03-06 38 10829 2022-01-13 38 10798 2021-12-26 46 11071 2022-05-05 46 11065 2022-05-01 46 10997 2022-04-03 78 11003 2022-04-06 78 10775 2021-12-12 78

Customerid OrderId OrderDate 1 11011 2016-04-09 1 10952 2016-03-16 1 10835 2016-01-15 2 10926 2016-03-04 2 10759 2015-11-28 2 10625 2015-08-08 3 10856 2016-01-28 3 10682 2015-09-25 3 10677 2015-09-22 4 11016 2016-04-10 4 10953 2016-03-16 4 10920 2016-03-03 5 10924 2016-03-04 5 10875 2016-02-06 5 10866 2016-02-03 6 11058 2016-04-29 6 10956 2016-03-17 6 10853 2016-01-27 7 10826 2016-01-12 7 10679 2015-09-23 7 10628 2015-08-12 8 10970 2016-03-24 8 10801 2015-12-29 8 10326 2014-10-10 9 11076 2016-05-06 9 10940 2016-03-11 9 10932 2016-03-06 10 11048 2016-04-24 10 11045 2016-04-23 10 11027 2016-04-16 11 11023 2016-04-14 11 10947 2016-03-13 11 10943 2016-03-11 12 11054 2016-04-28 12 10937 2016-03-10 12 10881 2016-02-11 13 10259 2014-07-18 14 11041 2016-04-22 14 11029 2016-04-16 14 10966 2016-03-20 15 11042 2016-04-22 15 10969 2016-03-23 15 10494 2015-04-02 16 10848 2016-01-23 16 10462 2015-03-03 16 10435 2015-02-04 17 11067 2016-05-04 17 11036 2016-04-20 17 10825 2016-01-09 18 10890 2016-02-16 18 10683 2015-09-26 18 10609 2015-07-24 19 11056 2016-04-28 19 11047 2016-04-24 19 11024 2016-04-15 20 11072 2016-05-05 20 11017 2016-04-13 20 11008 2016-04-08 21 10725 2015-10-31 21 10650 2015-08-29 21 10581 2015-06-26 23 10789 2015-12-22 23 10763 2015-12-03 23 10634 2015-08-15 24 11050 2016-04-27 24 11001 2016-04-06 24 10993 2016-04-01 25 11012 2016-04-09 25 10929 2016-03-05 25 10859 2016-01-29 26 10971 2016-03-24 26 10860 2016-01-29 26 10671 2015-09-17 27 11060 2016-04-30 27 11026 2016-04-15 27 10807 2015-12-31 28 10963 2016-03-19 28 10664 2015-09-10 28 10604 2015-07-18 29 10928 2016-03-05 29 10887 2016-02-13 29 10568 2015-06-13 30 11037 2016-04-21 30 11009 2016-04-08 30 10948 2016-03-13 31 11049 2016-04-24 31 10959 2016-03-18 31 10790 2015-12-22 32 11061 2016-04-30 32 11040 2016-04-22 32 11006 2016-04-07 33 10785 2015-12-18 33 10268 2014-07-30 34 11052 2016-04-27 34 11022 2016-04-14 34 10981 2016-03-27 35 11055 2016-04-28 35 10976 2016-03-25 35 10960 2016-03-19 36 10660 2015-09-08 36 10600 2015-07-16 36 10415 2015-01-15 37 11063 2016-04-30 37 10985 2016-03-30 37 10912 2016-02-26 38 10933 2016-03-06 38 10829 2016-01-13 38 10798 2015-12-26 39 11028 2016-04-16 39 10893 2016-02-18 39 10849 2016-01-23 40 10973 2016-03-24 40 10972 2016-03-24 40 10927 2016-03-05 41 11051 2016-04-27 41 10923 2016-03-03 41 10832 2016-01-14 42 10810 2016-01-01 42 10620 2015-08-05 42 10495 2015-04-03 43 10545 2015-05-22 43 10482 2015-03-21 44 11070 2016-05-05 44 10934 2016-03-09 44 10891 2016-02-17 45 10884 2016-02-12 45 10735 2015-11-10 45 10719 2015-10-27 46 11071 2016-05-05 46 11065 2016-05-01 46 10997 2016-04-03 47 11039 2016-04-21 47 11014 2016-04-10 47 10954 2016-03-17 48 11018 2016-04-13 48 10883 2016-02-12 48 10867 2016-02-03 49 10950 2016-03-16 49 10939 2016-03-10 49 10818 2016-01-07 50 11004 2016-04-07 50 10978 2016-03-26 50 10896 2016-02-19 51 10724 2015-10-30 51 10619 2015-08-04 51 10618 2015-08-01 52 10945 2016-03-12 52 10779 2015-12-16 52 10699 2015-10-09 53 11057 2016-04-29 53 10752 2015-11-24 53 10517 2015-04-24 54 10986 2016-03-30 54 10958 2016-03-18 54 10898 2016-02-20 55 11034 2016-04-20 55 10965 2016-03-20 55 10855 2016-01-27 56 11020 2016-04-14 56 10999 2016-04-03 56 10833 2016-01-15 58 11073 2016-05-05 58 10995 2016-04-02 58 10502 2015-04-10 59 11053 2016-04-27 59 10844 2016-01-21 59 10747 2015-11-19 60 11007 2016-04-08 60 10477 2015-03-17 60 10433 2015-02-03 61 10989 2016-03-31 61 10794 2015-12-24 61 10720 2015-10-28 62 11068 2016-05-04 62 10961 2016-03-19 62 10914 2016-02-27 63 11021 2016-04-14 63 10996 2016-04-02 63 10991 2016-04-01 64 11019 2016-04-13 64 10916 2016-02-27 64 10828 2016-01-13 65 11077 2016-05-06 65 11000 2016-04-06 65 10988 2016-03-31 66 11062 2016-04-30 66 11010 2016-04-09 66 10942 2016-03-11 67 11059 2016-04-29 67 10877 2016-02-09 67 10851 2016-01-26 68 11075 2016-05-06 68 11033 2016-04-17 68 10951 2016-03-16 69 11013 2016-04-09 69 10917 2016-03-02 69 10306 2014-09-16 70 11015 2016-04-10 70 10909 2016-02-26 70 10831 2016-01-1

**Preposition:** 

Retrieve the three most recent orders for each customer from the Sales.Customers table by using a cross apply operation with a subquery selecting the top three orders for each customer from the Sales.Orders table

**Tables:**

- Sales.Customers
- Sales.Orders

**Columns:**

- <span style="color: var(--vscode-foreground);">custid</span>
    
- orderid
    
- orderdate
    
- empid
    
- requireddate
    

**Predicate:**

- A cross apply operation to combine each row from the Sales.Customers table with the top three most recent orders for the corresponding customer from the Sales.Orders table
- The subquery takes the top three orders for each customer by filtering based on the customer ID (O.custid = C.custid), ordering the results by order date in descending order (latest orders first), and then by order ID in descending order (to break ties if multiple orders have the same order date)
- The OFFSET 0 ROWS FETCH NEXT 3 ROWS ONLY clause is used to limit the result set to the first three rows

In [93]:
--Query using TSQLV6 database
USE TSQLV6;

-- With OFFSET-FETCH
SELECT C.custid, A.orderid, A.orderdate
FROM Sales.Customers AS C
  CROSS APPLY
    (SELECT orderid, empid, orderdate, requireddate 
     FROM Sales.Orders AS O
     WHERE O.custid = C.custid
     ORDER BY orderdate DESC, orderid DESC
     OFFSET 0 ROWS FETCH NEXT 3 ROWS ONLY) AS A;


--Query using Northwinds2022TSQLV7
USE Northwinds2022TSQLV7;

-- With OFFSET-FETCH
SELECT C.CustomerId, A.OrderId, A.OrderDate
FROM Sales.[Customer] AS C
  CROSS APPLY
    (SELECT OrderId, EmployeeId, OrderDate, RequiredDate 
     FROM Sales.[Order] AS O
     WHERE O.CustomerId = C.CustomerId
     ORDER BY orderdate DESC, orderid DESC
     OFFSET 0 ROWS FETCH NEXT 3 ROWS ONLY) AS A;

(263 rows affected)

(263 rows affected)

Total execution time: 00:00:00.153

custid orderid orderdate 1 11011 2022-04-09 1 10952 2022-03-16 1 10835 2022-01-15 2 10926 2022-03-04 2 10759 2021-11-28 2 10625 2021-08-08 3 10856 2022-01-28 3 10682 2021-09-25 3 10677 2021-09-22 4 11016 2022-04-10 4 10953 2022-03-16 4 10920 2022-03-03 5 10924 2022-03-04 5 10875 2022-02-06 5 10866 2022-02-03 6 11058 2022-04-29 6 10956 2022-03-17 6 10853 2022-01-27 7 10826 2022-01-12 7 10679 2021-09-23 7 10628 2021-08-12 8 10970 2022-03-24 8 10801 2021-12-29 8 10326 2020-10-10 9 11076 2022-05-06 9 10940 2022-03-11 9 10932 2022-03-06 11 11023 2022-04-14 11 10947 2022-03-13 11 10943 2022-03-11 12 11054 2022-04-28 12 10937 2022-03-10 12 10881 2022-02-11 13 10259 2020-07-18 14 11041 2022-04-22 14 11029 2022-04-16 14 10966 2022-03-20 16 10848 2022-01-23 16 10462 2021-03-03 16 10435 2021-02-04 17 11067 2022-05-04 17 11036 2022-04-20 17 10825 2022-01-09 18 10890 2022-02-16 18 10683 2021-09-26 18 10609 2021-07-24 19 11056 2022-04-28 19 11047 2022-04-24 19 11024 2022-04-15 20 11072 2022-05-05 20 11017 2022-04-13 20 11008 2022-04-08 23 10789 2021-12-22 23 10763 2021-12-03 23 10634 2021-08-15 24 11050 2022-04-27 24 11001 2022-04-06 24 10993 2022-04-01 25 11012 2022-04-09 25 10929 2022-03-05 25 10859 2022-01-29 26 10971 2022-03-24 26 10860 2022-01-29 26 10671 2021-09-17 27 11060 2022-04-30 27 11026 2022-04-15 27 10807 2021-12-31 28 10963 2022-03-19 28 10664 2021-09-10 28 10604 2021-07-18 29 10928 2022-03-05 29 10887 2022-02-13 29 10568 2021-06-13 30 11037 2022-04-21 30 11009 2022-04-08 30 10948 2022-03-13 39 11028 2022-04-16 39 10893 2022-02-18 39 10849 2022-01-23 40 10973 2022-03-24 40 10972 2022-03-24 40 10927 2022-03-05 41 11051 2022-04-27 41 10923 2022-03-03 41 10832 2022-01-14 44 11070 2022-05-05 44 10934 2022-03-09 44 10891 2022-02-17 49 10950 2022-03-16 49 10939 2022-03-10 49 10818 2022-01-07 50 11004 2022-04-07 50 10978 2022-03-26 50 10896 2022-02-19 52 10945 2022-03-12 52 10779 2021-12-16 52 10699 2021-10-09 53 11057 2022-04-29 53 10752 2021-11-24 53 10517 2021-04-24 54 10986 2022-03-30 54 10958 2022-03-18 54 10898 2022-02-20 56 11020 2022-04-14 56 10999 2022-04-03 56 10833 2022-01-15 58 11073 2022-05-05 58 10995 2022-04-02 58 10502 2021-04-10 59 11053 2022-04-27 59 10844 2022-01-21 59 10747 2021-11-19 60 11007 2022-04-08 60 10477 2021-03-17 60 10433 2021-02-03 63 11021 2022-04-14 63 10996 2022-04-02 63 10991 2022-04-01 64 11019 2022-04-13 64 10916 2022-02-27 64 10828 2022-01-13 66 11062 2022-04-30 66 11010 2022-04-09 66 10942 2022-03-11 68 11075 2022-05-06 68 11033 2022-04-17 68 10951 2022-03-16 69 11013 2022-04-09 69 10917 2022-03-02 69 10306 2020-09-16 70 11015 2022-04-10 70 10909 2022-02-26 70 10831 2022-01-14 72 10869 2022-02-04 72 10804 2021-12-30 72 10800 2021-12-26 73 11074 2022-05-06 73 10802 2021-12-29 73 10669 2021-09-15 74 11043 2022-04-22 74 10964 2022-03-20 74 10907 2022-02-25 76 11038 2022-04-21 76 11035 2022-04-20 76 10930 2022-03-06 79 10967 2022-03-23 79 10608 2021-07-23 79 10548 2021-05-26 80 11069 2022-05-04 80 10915 2022-02-27 80 10842 2022-01-20 83 10994 2022-04-02 83 10946 2022-03-12 83 10921 2022-03-03 84 10850 2022-01-23 84 10843 2022-01-21 84 10814 2022-01-05 85 10739 2021-11-12 85 10737 2021-11-11 85 10295 2020-09-02 86 11046 2022-04-23 86 10668 2021-09-15 86 10651 2021-09-01 87 11025 2022-04-15 87 10781 2021-12-17 87 10750 2021-11-21 90 11005 2022-04-07 90 10910 2022-02-26 90 10879 2022-02-10 91 11044 2022-04-23 91 10998 2022-04-03 91 10906 2022-02-25 55 11034 2022-04-20 55 10965 2022-03-20 55 10855 2022-01-27 10 11048 2022-04-24 10 11045 2022-04-23 10 11027 2022-04-16 42 10810 2022-01-01 42 10620 2021-08-05 42 10495 2021-04-03 45 10884 2022-02-12 45 10735 2021-11-10 45 10719 2021-10-27 37 11063 2022-04-30 37 10985 2022-03-30 37 10912 2022-02-26 33 10785 2021-12-18 33 10268 2020-07-30 71 11064 2022-05-01 71 11031 2022-04-17 71 11030 2022-04-17 38 10933 2022-03-06 38 10829 2022-01-13 38 10798 2021-12-26 46 11071 2022-05-05 46 11065 2022-05-01 46 10997 2022-04-03 78 11003 2022-04-06 78 10775 2021-12-12 78

CustomerId OrderId OrderDate 1 11011 2016-04-09 1 10952 2016-03-16 1 10835 2016-01-15 2 10926 2016-03-04 2 10759 2015-11-28 2 10625 2015-08-08 3 10856 2016-01-28 3 10682 2015-09-25 3 10677 2015-09-22 4 11016 2016-04-10 4 10953 2016-03-16 4 10920 2016-03-03 5 10924 2016-03-04 5 10875 2016-02-06 5 10866 2016-02-03 6 11058 2016-04-29 6 10956 2016-03-17 6 10853 2016-01-27 7 10826 2016-01-12 7 10679 2015-09-23 7 10628 2015-08-12 8 10970 2016-03-24 8 10801 2015-12-29 8 10326 2014-10-10 9 11076 2016-05-06 9 10940 2016-03-11 9 10932 2016-03-06 10 11048 2016-04-24 10 11045 2016-04-23 10 11027 2016-04-16 11 11023 2016-04-14 11 10947 2016-03-13 11 10943 2016-03-11 12 11054 2016-04-28 12 10937 2016-03-10 12 10881 2016-02-11 13 10259 2014-07-18 14 11041 2016-04-22 14 11029 2016-04-16 14 10966 2016-03-20 15 11042 2016-04-22 15 10969 2016-03-23 15 10494 2015-04-02 16 10848 2016-01-23 16 10462 2015-03-03 16 10435 2015-02-04 17 11067 2016-05-04 17 11036 2016-04-20 17 10825 2016-01-09 18 10890 2016-02-16 18 10683 2015-09-26 18 10609 2015-07-24 19 11056 2016-04-28 19 11047 2016-04-24 19 11024 2016-04-15 20 11072 2016-05-05 20 11017 2016-04-13 20 11008 2016-04-08 21 10725 2015-10-31 21 10650 2015-08-29 21 10581 2015-06-26 23 10789 2015-12-22 23 10763 2015-12-03 23 10634 2015-08-15 24 11050 2016-04-27 24 11001 2016-04-06 24 10993 2016-04-01 25 11012 2016-04-09 25 10929 2016-03-05 25 10859 2016-01-29 26 10971 2016-03-24 26 10860 2016-01-29 26 10671 2015-09-17 27 11060 2016-04-30 27 11026 2016-04-15 27 10807 2015-12-31 28 10963 2016-03-19 28 10664 2015-09-10 28 10604 2015-07-18 29 10928 2016-03-05 29 10887 2016-02-13 29 10568 2015-06-13 30 11037 2016-04-21 30 11009 2016-04-08 30 10948 2016-03-13 31 11049 2016-04-24 31 10959 2016-03-18 31 10790 2015-12-22 32 11061 2016-04-30 32 11040 2016-04-22 32 11006 2016-04-07 33 10785 2015-12-18 33 10268 2014-07-30 34 11052 2016-04-27 34 11022 2016-04-14 34 10981 2016-03-27 35 11055 2016-04-28 35 10976 2016-03-25 35 10960 2016-03-19 36 10660 2015-09-08 36 10600 2015-07-16 36 10415 2015-01-15 37 11063 2016-04-30 37 10985 2016-03-30 37 10912 2016-02-26 38 10933 2016-03-06 38 10829 2016-01-13 38 10798 2015-12-26 39 11028 2016-04-16 39 10893 2016-02-18 39 10849 2016-01-23 40 10973 2016-03-24 40 10972 2016-03-24 40 10927 2016-03-05 41 11051 2016-04-27 41 10923 2016-03-03 41 10832 2016-01-14 42 10810 2016-01-01 42 10620 2015-08-05 42 10495 2015-04-03 43 10545 2015-05-22 43 10482 2015-03-21 44 11070 2016-05-05 44 10934 2016-03-09 44 10891 2016-02-17 45 10884 2016-02-12 45 10735 2015-11-10 45 10719 2015-10-27 46 11071 2016-05-05 46 11065 2016-05-01 46 10997 2016-04-03 47 11039 2016-04-21 47 11014 2016-04-10 47 10954 2016-03-17 48 11018 2016-04-13 48 10883 2016-02-12 48 10867 2016-02-03 49 10950 2016-03-16 49 10939 2016-03-10 49 10818 2016-01-07 50 11004 2016-04-07 50 10978 2016-03-26 50 10896 2016-02-19 51 10724 2015-10-30 51 10619 2015-08-04 51 10618 2015-08-01 52 10945 2016-03-12 52 10779 2015-12-16 52 10699 2015-10-09 53 11057 2016-04-29 53 10752 2015-11-24 53 10517 2015-04-24 54 10986 2016-03-30 54 10958 2016-03-18 54 10898 2016-02-20 55 11034 2016-04-20 55 10965 2016-03-20 55 10855 2016-01-27 56 11020 2016-04-14 56 10999 2016-04-03 56 10833 2016-01-15 58 11073 2016-05-05 58 10995 2016-04-02 58 10502 2015-04-10 59 11053 2016-04-27 59 10844 2016-01-21 59 10747 2015-11-19 60 11007 2016-04-08 60 10477 2015-03-17 60 10433 2015-02-03 61 10989 2016-03-31 61 10794 2015-12-24 61 10720 2015-10-28 62 11068 2016-05-04 62 10961 2016-03-19 62 10914 2016-02-27 63 11021 2016-04-14 63 10996 2016-04-02 63 10991 2016-04-01 64 11019 2016-04-13 64 10916 2016-02-27 64 10828 2016-01-13 65 11077 2016-05-06 65 11000 2016-04-06 65 10988 2016-03-31 66 11062 2016-04-30 66 11010 2016-04-09 66 10942 2016-03-11 67 11059 2016-04-29 67 10877 2016-02-09 67 10851 2016-01-26 68 11075 2016-05-06 68 11033 2016-04-17 68 10951 2016-03-16 69 11013 2016-04-09 69 10917 2016-03-02 69 10306 2014-09-16 70 11015 2016-04-10 70 10909 2016-02-26 70 10831 2016-01-1

**Preposition:** 

Retrieve the three most recent orders for each customer from the Sales.Customers table while preserving all customers, including those without any orders, by performing an outer apply operation with a subquery selecting the top three orders for each customer from the Sales.Orders table

**Tables:**

- Sales.Customers
- Sales.Orders

**Columns:**

- <span style="color: var(--vscode-foreground);">custid</span>
    
- orderid
    
- orderdate
    
- empid
    
- requireddate
    

**Predicate:**

- A outter apply operation is used to combine each row from the Sales.Customers table with the top three most recent orders for the corresponding customer from the Sales.Orders table
- The subquery takes the top three orders for each customer by filtering based on the customer ID (O.custid = C.custid), ordering the results by order date in descending order (latest orders first), and then by order ID in descending order (to break ties if multiple orders have the same order date)

In [95]:
--Query using TSQLV6 database
USE TSQLV6;

-- 3 most recent orders for each customer, preserve all customers
SELECT C.custid, A.orderid, A.orderdate
FROM Sales.Customers AS C
  OUTER APPLY
    (SELECT TOP (3) orderid, empid, orderdate, requireddate 
     FROM Sales.Orders AS O
     WHERE O.custid = C.custid
     ORDER BY orderdate DESC, orderid DESC) AS A;
GO

--Query using Northwinds2022TSQLV7
USE Northwinds2022TSQLV7;

-- 3 most recent orders for each customer, preserve all customers
SELECT C.CustomerId, A.OrderId, A.OrderDate
FROM Sales.[Customer] AS C
  OUTER APPLY
    (SELECT TOP (3) OrderId, EmployeeId, OrderDate, RequiredDate 
     FROM Sales.[Order] AS O
     WHERE O.CustomerId = C.CustomerId
     ORDER BY orderdate DESC, orderid DESC) AS A;
GO

(265 rows affected)

(265 rows affected)

Total execution time: 00:00:00.100

custid orderid orderdate 1 11011 2022-04-09 1 10952 2022-03-16 1 10835 2022-01-15 2 10926 2022-03-04 2 10759 2021-11-28 2 10625 2021-08-08 3 10856 2022-01-28 3 10682 2021-09-25 3 10677 2021-09-22 4 11016 2022-04-10 4 10953 2022-03-16 4 10920 2022-03-03 5 10924 2022-03-04 5 10875 2022-02-06 5 10866 2022-02-03 6 11058 2022-04-29 6 10956 2022-03-17 6 10853 2022-01-27 7 10826 2022-01-12 7 10679 2021-09-23 7 10628 2021-08-12 8 10970 2022-03-24 8 10801 2021-12-29 8 10326 2020-10-10 9 11076 2022-05-06 9 10940 2022-03-11 9 10932 2022-03-06 11 11023 2022-04-14 11 10947 2022-03-13 11 10943 2022-03-11 12 11054 2022-04-28 12 10937 2022-03-10 12 10881 2022-02-11 13 10259 2020-07-18 14 11041 2022-04-22 14 11029 2022-04-16 14 10966 2022-03-20 16 10848 2022-01-23 16 10462 2021-03-03 16 10435 2021-02-04 17 11067 2022-05-04 17 11036 2022-04-20 17 10825 2022-01-09 18 10890 2022-02-16 18 10683 2021-09-26 18 10609 2021-07-24 19 11056 2022-04-28 19 11047 2022-04-24 19 11024 2022-04-15 20 11072 2022-05-05 20 11017 2022-04-13 20 11008 2022-04-08 22 NULL NULL 23 10789 2021-12-22 23 10763 2021-12-03 23 10634 2021-08-15 24 11050 2022-04-27 24 11001 2022-04-06 24 10993 2022-04-01 25 11012 2022-04-09 25 10929 2022-03-05 25 10859 2022-01-29 26 10971 2022-03-24 26 10860 2022-01-29 26 10671 2021-09-17 27 11060 2022-04-30 27 11026 2022-04-15 27 10807 2021-12-31 28 10963 2022-03-19 28 10664 2021-09-10 28 10604 2021-07-18 29 10928 2022-03-05 29 10887 2022-02-13 29 10568 2021-06-13 30 11037 2022-04-21 30 11009 2022-04-08 30 10948 2022-03-13 39 11028 2022-04-16 39 10893 2022-02-18 39 10849 2022-01-23 40 10973 2022-03-24 40 10972 2022-03-24 40 10927 2022-03-05 41 11051 2022-04-27 41 10923 2022-03-03 41 10832 2022-01-14 44 11070 2022-05-05 44 10934 2022-03-09 44 10891 2022-02-17 49 10950 2022-03-16 49 10939 2022-03-10 49 10818 2022-01-07 50 11004 2022-04-07 50 10978 2022-03-26 50 10896 2022-02-19 52 10945 2022-03-12 52 10779 2021-12-16 52 10699 2021-10-09 53 11057 2022-04-29 53 10752 2021-11-24 53 10517 2021-04-24 54 10986 2022-03-30 54 10958 2022-03-18 54 10898 2022-02-20 56 11020 2022-04-14 56 10999 2022-04-03 56 10833 2022-01-15 57 NULL NULL 58 11073 2022-05-05 58 10995 2022-04-02 58 10502 2021-04-10 59 11053 2022-04-27 59 10844 2022-01-21 59 10747 2021-11-19 60 11007 2022-04-08 60 10477 2021-03-17 60 10433 2021-02-03 63 11021 2022-04-14 63 10996 2022-04-02 63 10991 2022-04-01 64 11019 2022-04-13 64 10916 2022-02-27 64 10828 2022-01-13 66 11062 2022-04-30 66 11010 2022-04-09 66 10942 2022-03-11 68 11075 2022-05-06 68 11033 2022-04-17 68 10951 2022-03-16 69 11013 2022-04-09 69 10917 2022-03-02 69 10306 2020-09-16 70 11015 2022-04-10 70 10909 2022-02-26 70 10831 2022-01-14 72 10869 2022-02-04 72 10804 2021-12-30 72 10800 2021-12-26 73 11074 2022-05-06 73 10802 2021-12-29 73 10669 2021-09-15 74 11043 2022-04-22 74 10964 2022-03-20 74 10907 2022-02-25 76 11038 2022-04-21 76 11035 2022-04-20 76 10930 2022-03-06 79 10967 2022-03-23 79 10608 2021-07-23 79 10548 2021-05-26 80 11069 2022-05-04 80 10915 2022-02-27 80 10842 2022-01-20 83 10994 2022-04-02 83 10946 2022-03-12 83 10921 2022-03-03 84 10850 2022-01-23 84 10843 2022-01-21 84 10814 2022-01-05 85 10739 2021-11-12 85 10737 2021-11-11 85 10295 2020-09-02 86 11046 2022-04-23 86 10668 2021-09-15 86 10651 2021-09-01 87 11025 2022-04-15 87 10781 2021-12-17 87 10750 2021-11-21 90 11005 2022-04-07 90 10910 2022-02-26 90 10879 2022-02-10 91 11044 2022-04-23 91 10998 2022-04-03 91 10906 2022-02-25 55 11034 2022-04-20 55 10965 2022-03-20 55 10855 2022-01-27 10 11048 2022-04-24 10 11045 2022-04-23 10 11027 2022-04-16 42 10810 2022-01-01 42 10620 2021-08-05 42 10495 2021-04-03 45 10884 2022-02-12 45 10735 2021-11-10 45 10719 2021-10-27 37 11063 2022-04-30 37 10985 2022-03-30 37 10912 2022-02-26 33 10785 2021-12-18 33 10268 2020-07-30 71 11064 2022-05-01 71 11031 2022-04-17 71 11030 2022-04-17 38 10933 2022-03-06 38 10829 2022-01-13 38 10798 2021-12-26 46 11071 2022-05-05 46 11065 2022-05-01 46 10997 2022-04-03 78 11003 2022-04

CustomerId OrderId OrderDate 1 11011 2016-04-09 1 10952 2016-03-16 1 10835 2016-01-15 2 10926 2016-03-04 2 10759 2015-11-28 2 10625 2015-08-08 3 10856 2016-01-28 3 10682 2015-09-25 3 10677 2015-09-22 4 11016 2016-04-10 4 10953 2016-03-16 4 10920 2016-03-03 5 10924 2016-03-04 5 10875 2016-02-06 5 10866 2016-02-03 6 11058 2016-04-29 6 10956 2016-03-17 6 10853 2016-01-27 7 10826 2016-01-12 7 10679 2015-09-23 7 10628 2015-08-12 8 10970 2016-03-24 8 10801 2015-12-29 8 10326 2014-10-10 9 11076 2016-05-06 9 10940 2016-03-11 9 10932 2016-03-06 10 11048 2016-04-24 10 11045 2016-04-23 10 11027 2016-04-16 11 11023 2016-04-14 11 10947 2016-03-13 11 10943 2016-03-11 12 11054 2016-04-28 12 10937 2016-03-10 12 10881 2016-02-11 13 10259 2014-07-18 14 11041 2016-04-22 14 11029 2016-04-16 14 10966 2016-03-20 15 11042 2016-04-22 15 10969 2016-03-23 15 10494 2015-04-02 16 10848 2016-01-23 16 10462 2015-03-03 16 10435 2015-02-04 17 11067 2016-05-04 17 11036 2016-04-20 17 10825 2016-01-09 18 10890 2016-02-16 18 10683 2015-09-26 18 10609 2015-07-24 19 11056 2016-04-28 19 11047 2016-04-24 19 11024 2016-04-15 20 11072 2016-05-05 20 11017 2016-04-13 20 11008 2016-04-08 21 10725 2015-10-31 21 10650 2015-08-29 21 10581 2015-06-26 22 NULL NULL 23 10789 2015-12-22 23 10763 2015-12-03 23 10634 2015-08-15 24 11050 2016-04-27 24 11001 2016-04-06 24 10993 2016-04-01 25 11012 2016-04-09 25 10929 2016-03-05 25 10859 2016-01-29 26 10971 2016-03-24 26 10860 2016-01-29 26 10671 2015-09-17 27 11060 2016-04-30 27 11026 2016-04-15 27 10807 2015-12-31 28 10963 2016-03-19 28 10664 2015-09-10 28 10604 2015-07-18 29 10928 2016-03-05 29 10887 2016-02-13 29 10568 2015-06-13 30 11037 2016-04-21 30 11009 2016-04-08 30 10948 2016-03-13 31 11049 2016-04-24 31 10959 2016-03-18 31 10790 2015-12-22 32 11061 2016-04-30 32 11040 2016-04-22 32 11006 2016-04-07 33 10785 2015-12-18 33 10268 2014-07-30 34 11052 2016-04-27 34 11022 2016-04-14 34 10981 2016-03-27 35 11055 2016-04-28 35 10976 2016-03-25 35 10960 2016-03-19 36 10660 2015-09-08 36 10600 2015-07-16 36 10415 2015-01-15 37 11063 2016-04-30 37 10985 2016-03-30 37 10912 2016-02-26 38 10933 2016-03-06 38 10829 2016-01-13 38 10798 2015-12-26 39 11028 2016-04-16 39 10893 2016-02-18 39 10849 2016-01-23 40 10973 2016-03-24 40 10972 2016-03-24 40 10927 2016-03-05 41 11051 2016-04-27 41 10923 2016-03-03 41 10832 2016-01-14 42 10810 2016-01-01 42 10620 2015-08-05 42 10495 2015-04-03 43 10545 2015-05-22 43 10482 2015-03-21 44 11070 2016-05-05 44 10934 2016-03-09 44 10891 2016-02-17 45 10884 2016-02-12 45 10735 2015-11-10 45 10719 2015-10-27 46 11071 2016-05-05 46 11065 2016-05-01 46 10997 2016-04-03 47 11039 2016-04-21 47 11014 2016-04-10 47 10954 2016-03-17 48 11018 2016-04-13 48 10883 2016-02-12 48 10867 2016-02-03 49 10950 2016-03-16 49 10939 2016-03-10 49 10818 2016-01-07 50 11004 2016-04-07 50 10978 2016-03-26 50 10896 2016-02-19 51 10724 2015-10-30 51 10619 2015-08-04 51 10618 2015-08-01 52 10945 2016-03-12 52 10779 2015-12-16 52 10699 2015-10-09 53 11057 2016-04-29 53 10752 2015-11-24 53 10517 2015-04-24 54 10986 2016-03-30 54 10958 2016-03-18 54 10898 2016-02-20 55 11034 2016-04-20 55 10965 2016-03-20 55 10855 2016-01-27 56 11020 2016-04-14 56 10999 2016-04-03 56 10833 2016-01-15 57 NULL NULL 58 11073 2016-05-05 58 10995 2016-04-02 58 10502 2015-04-10 59 11053 2016-04-27 59 10844 2016-01-21 59 10747 2015-11-19 60 11007 2016-04-08 60 10477 2015-03-17 60 10433 2015-02-03 61 10989 2016-03-31 61 10794 2015-12-24 61 10720 2015-10-28 62 11068 2016-05-04 62 10961 2016-03-19 62 10914 2016-02-27 63 11021 2016-04-14 63 10996 2016-04-02 63 10991 2016-04-01 64 11019 2016-04-13 64 10916 2016-02-27 64 10828 2016-01-13 65 11077 2016-05-06 65 11000 2016-04-06 65 10988 2016-03-31 66 11062 2016-04-30 66 11010 2016-04-09 66 10942 2016-03-11 67 11059 2016-04-29 67 10877 2016-02-09 67 10851 2016-01-26 68 11075 2016-05-06 68 11033 2016-04-17 68 10951 2016-03-16 69 11013 2016-04-09 69 10917 2016-03-02 69 10306 2014-09-16 70 11015 2016-04-10 70 10909 201

**Preposition:** 

Create or alter a table named dbo.TopOrders and take two parameters @custid and @n, and returns a table containing the top @n orders for a specific customer ID (@custid). Then use CROSS APPLY to apply the function to each row in the Sales.Customers table, retrieving the top 3 orders for each customer

**Tables:**

- Sales.Orders
- Sales.Customers

**Columns:**

- orderid
- empid
- orderdate
- requireddate
- custid
- companyname

**Predicate:**

- This function selects the data from Sales.Customers table and retrieves the top 3 orders for each customer based on the specified parameters (@custid and 3)
- It is then filtered based on customer ID, ordering them by order date in descending order and then by order ID in descending order

In [19]:
--Query using TSQLV6 database
USE TSQLV6;
GO
-- Creation Script for the Function TopOrders
CREATE OR ALTER FUNCTION dbo.TopOrders
  (@n AS INT, @custid AS INT)
  RETURNS TABLE
AS
RETURN
  SELECT TOP (@n) orderid, empid, orderdate, requireddate
  FROM Sales.Orders
  WHERE custid = @custid
  ORDER BY orderdate DESC, orderid DESC;
GO

SELECT
  C.custid, C.companyname,
  A.orderid, A.empid, A.orderdate, A.requireddate 
FROM Sales.Customers AS C
  CROSS APPLY dbo.TopOrders(C.custid, 3) AS A;


--Query using Northwinds2022TSQLV7
USE Northwinds2022TSQLV7;
GO
-- Creation Script for the Function TopOrders
CREATE OR ALTER FUNCTION dbo.TopOrders
  (@custid AS INT, @n AS INT)
  RETURNS TABLE
AS
RETURN
  SELECT TOP (@n) OrderId, EmployeeId, OrderDate, RequiredDate
  FROM Sales.[Order]
  WHERE CustomerId = @custid
  ORDER BY OrderDate DESC, OrderId DESC;
GO

SELECT
  C.CustomerId, C.CustomerCompanyName,
  A.OrderId, A.EmployeeId, A.OrderDate, A.RequiredDate 
FROM Sales.[Customer] AS C
  CROSS APPLY dbo.TopOrders(C.CustomerId, 3) AS A;


Commands completed successfully.

Commands completed successfully.

(616 rows affected)

Commands completed successfully.

(263 rows affected)

Total execution time: 00:00:00.077

custid companyname orderid empid orderdate requireddate 72 Customer AHPOP 10856 3 2022-01-28 2022-02-25 72 Customer AHPOP 10682 3 2021-09-25 2021-10-23 72 Customer AHPOP 10677 1 2021-09-22 2021-10-20 72 Customer AHPOP 10573 7 2021-06-19 2021-07-17 72 Customer AHPOP 10535 4 2021-05-13 2021-06-10 72 Customer AHPOP 10507 7 2021-04-15 2021-05-13 72 Customer AHPOP 10365 3 2020-11-27 2020-12-25 58 Customer AHXHT 10856 3 2022-01-28 2022-02-25 58 Customer AHXHT 10682 3 2021-09-25 2021-10-23 58 Customer AHXHT 10677 1 2021-09-22 2021-10-20 58 Customer AHXHT 10573 7 2021-06-19 2021-07-17 58 Customer AHXHT 10535 4 2021-05-13 2021-06-10 58 Customer AHXHT 10507 7 2021-04-15 2021-05-13 58 Customer AHXHT 10365 3 2020-11-27 2020-12-25 25 Customer AZJED 10856 3 2022-01-28 2022-02-25 25 Customer AZJED 10682 3 2021-09-25 2021-10-23 25 Customer AZJED 10677 1 2021-09-22 2021-10-20 25 Customer AZJED 10573 7 2021-06-19 2021-07-17 25 Customer AZJED 10535 4 2021-05-13 2021-06-10 25 Customer AZJED 10507 7 2021-04-15 2021-05-13 25 Customer AZJED 10365 3 2020-11-27 2020-12-25 18 Customer BSVAR 10856 3 2022-01-28 2022-02-25 18 Customer BSVAR 10682 3 2021-09-25 2021-10-23 18 Customer BSVAR 10677 1 2021-09-22 2021-10-20 18 Customer BSVAR 10573 7 2021-06-19 2021-07-17 18 Customer BSVAR 10535 4 2021-05-13 2021-06-10 18 Customer BSVAR 10507 7 2021-04-15 2021-05-13 18 Customer BSVAR 10365 3 2020-11-27 2020-12-25 91 Customer CCFIZ 10856 3 2022-01-28 2022-02-25 91 Customer CCFIZ 10682 3 2021-09-25 2021-10-23 91 Customer CCFIZ 10677 1 2021-09-22 2021-10-20 91 Customer CCFIZ 10573 7 2021-06-19 2021-07-17 91 Customer CCFIZ 10535 4 2021-05-13 2021-06-10 91 Customer CCFIZ 10507 7 2021-04-15 2021-05-13 91 Customer CCFIZ 10365 3 2020-11-27 2020-12-25 68 Customer CCKOT 10856 3 2022-01-28 2022-02-25 68 Customer CCKOT 10682 3 2021-09-25 2021-10-23 68 Customer CCKOT 10677 1 2021-09-22 2021-10-20 68 Customer CCKOT 10573 7 2021-06-19 2021-07-17 68 Customer CCKOT 10535 4 2021-05-13 2021-06-10 68 Customer CCKOT 10507 7 2021-04-15 2021-05-13 68 Customer CCKOT 10365 3 2020-11-27 2020-12-25 49 Customer CQRAA 10856 3 2022-01-28 2022-02-25 49 Customer CQRAA 10682 3 2021-09-25 2021-10-23 49 Customer CQRAA 10677 1 2021-09-22 2021-10-20 49 Customer CQRAA 10573 7 2021-06-19 2021-07-17 49 Customer CQRAA 10535 4 2021-05-13 2021-06-10 49 Customer CQRAA 10507 7 2021-04-15 2021-05-13 49 Customer CQRAA 10365 3 2020-11-27 2020-12-25 24 Customer CYZTN 10856 3 2022-01-28 2022-02-25 24 Customer CYZTN 10682 3 2021-09-25 2021-10-23 24 Customer CYZTN 10677 1 2021-09-22 2021-10-20 24 Customer CYZTN 10573 7 2021-06-19 2021-07-17 24 Customer CYZTN 10535 4 2021-05-13 2021-06-10 24 Customer CYZTN 10507 7 2021-04-15 2021-05-13 24 Customer CYZTN 10365 3 2020-11-27 2020-12-25 22 Customer DTDMN 10856 3 2022-01-28 2022-02-25 22 Customer DTDMN 10682 3 2021-09-25 2021-10-23 22 Customer DTDMN 10677 1 2021-09-22 2021-10-20 22 Customer DTDMN 10573 7 2021-06-19 2021-07-17 22 Customer DTDMN 10535 4 2021-05-13 2021-06-10 22 Customer DTDMN 10507 7 2021-04-15 2021-05-13 22 Customer DTDMN 10365 3 2020-11-27 2020-12-25 48 Customer DVFMB 10856 3 2022-01-28 2022-02-25 48 Customer DVFMB 10682 3 2021-09-25 2021-10-23 48 Customer DVFMB 10677 1 2021-09-22 2021-10-20 48 Customer DVFMB 10573 7 2021-06-19 2021-07-17 48 Customer DVFMB 10535 4 2021-05-13 2021-06-10 48 Customer DVFMB 10507 7 2021-04-15 2021-05-13 48 Customer DVFMB 10365 3 2020-11-27 2020-12-25 10 Customer EEALV 10856 3 2022-01-28 2022-02-25 10 Customer EEALV 10682 3 2021-09-25 2021-10-23 10 Customer EEALV 10677 1 2021-09-22 2021-10-20 10 Customer EEALV 10573 7 2021-06-19 2021-07-17 10 Customer EEALV 10535 4 2021-05-13 2021-06-10 10 Customer EEALV 10507 7 2021-04-15 2021-05-13 10 Customer EEALV 10365 3 2020-11-27 2020-12-25 40 Customer EFFTC 10856 3 2022-01-28 2022-02-25 40 Customer EFFTC 10682 3 2021-09-25 2021-10-23 40 Customer EFFTC 10677 1 2021-09-22 2021-10-20 40 Customer EFFTC 10573 7 2021-06-19 2021-07-17 40 Customer EFFTC 10535 4 2021-05-13 2021-06-10 40 Custo

CustomerId CustomerCompanyName OrderId EmployeeId OrderDate RequiredDate 1 Customer NRZBB 11011 3 2016-04-09 2016-05-07 1 Customer NRZBB 10952 1 2016-03-16 2016-04-27 1 Customer NRZBB 10835 1 2016-01-15 2016-02-12 2 Customer MLTDN 10926 4 2016-03-04 2016-04-01 2 Customer MLTDN 10759 3 2015-11-28 2015-12-26 2 Customer MLTDN 10625 3 2015-08-08 2015-09-05 3 Customer KBUDE 10856 3 2016-01-28 2016-02-25 3 Customer KBUDE 10682 3 2015-09-25 2015-10-23 3 Customer KBUDE 10677 1 2015-09-22 2015-10-20 4 Customer HFBZG 11016 9 2016-04-10 2016-05-08 4 Customer HFBZG 10953 9 2016-03-16 2016-03-30 4 Customer HFBZG 10920 4 2016-03-03 2016-03-31 5 Customer HGVLZ 10924 3 2016-03-04 2016-04-01 5 Customer HGVLZ 10875 4 2016-02-06 2016-03-06 5 Customer HGVLZ 10866 5 2016-02-03 2016-03-03 6 Customer XHXJV 11058 9 2016-04-29 2016-05-27 6 Customer XHXJV 10956 6 2016-03-17 2016-04-28 6 Customer XHXJV 10853 9 2016-01-27 2016-02-24 7 Customer QXVLA 10826 6 2016-01-12 2016-02-09 7 Customer QXVLA 10679 8 2015-09-23 2015-10-21 7 Customer QXVLA 10628 4 2015-08-12 2015-09-09 8 Customer QUHWH 10970 9 2016-03-24 2016-04-07 8 Customer QUHWH 10801 4 2015-12-29 2016-01-26 8 Customer QUHWH 10326 4 2014-10-10 2014-11-07 9 Customer RTXGC 11076 4 2016-05-06 2016-06-03 9 Customer RTXGC 10940 8 2016-03-11 2016-04-08 9 Customer RTXGC 10932 8 2016-03-06 2016-04-03 10 Customer EEALV 11048 7 2016-04-24 2016-05-22 10 Customer EEALV 11045 6 2016-04-23 2016-05-21 10 Customer EEALV 11027 1 2016-04-16 2016-05-14 11 Customer UBHAU 11023 1 2016-04-14 2016-04-28 11 Customer UBHAU 10947 3 2016-03-13 2016-04-10 11 Customer UBHAU 10943 4 2016-03-11 2016-04-08 12 Customer PSNMQ 11054 8 2016-04-28 2016-05-26 12 Customer PSNMQ 10937 7 2016-03-10 2016-03-24 12 Customer PSNMQ 10881 4 2016-02-11 2016-03-11 13 Customer VMLOG 10259 4 2014-07-18 2014-08-15 14 Customer WNMAF 11041 3 2016-04-22 2016-05-20 14 Customer WNMAF 11029 4 2016-04-16 2016-05-14 14 Customer WNMAF 10966 4 2016-03-20 2016-04-17 15 Customer JUWXK 11042 2 2016-04-22 2016-05-06 15 Customer JUWXK 10969 1 2016-03-23 2016-04-20 15 Customer JUWXK 10494 4 2015-04-02 2015-04-30 16 Customer GYBBY 10848 7 2016-01-23 2016-02-20 16 Customer GYBBY 10462 2 2015-03-03 2015-03-31 16 Customer GYBBY 10435 8 2015-02-04 2015-03-18 17 Customer FEVNN 11067 1 2016-05-04 2016-05-18 17 Customer FEVNN 11036 8 2016-04-20 2016-05-18 17 Customer FEVNN 10825 1 2016-01-09 2016-02-06 18 Customer BSVAR 10890 7 2016-02-16 2016-03-16 18 Customer BSVAR 10683 2 2015-09-26 2015-10-24 18 Customer BSVAR 10609 7 2015-07-24 2015-08-21 19 Customer RFNQC 11056 8 2016-04-28 2016-05-12 19 Customer RFNQC 11047 7 2016-04-24 2016-05-22 19 Customer RFNQC 11024 4 2016-04-15 2016-05-13 20 Customer THHDP 11072 4 2016-05-05 2016-06-02 20 Customer THHDP 11017 9 2016-04-13 2016-05-11 20 Customer THHDP 11008 7 2016-04-08 2016-05-06 21 Customer KIDPX 10725 4 2015-10-31 2015-11-28 21 Customer KIDPX 10650 5 2015-08-29 2015-09-26 21 Customer KIDPX 10581 3 2015-06-26 2015-07-24 23 Customer WVFAF 10789 1 2015-12-22 2016-01-19 23 Customer WVFAF 10763 3 2015-12-03 2015-12-31 23 Customer WVFAF 10634 4 2015-08-15 2015-09-12 24 Customer CYZTN 11050 8 2016-04-27 2016-05-25 24 Customer CYZTN 11001 2 2016-04-06 2016-05-04 24 Customer CYZTN 10993 7 2016-04-01 2016-04-29 25 Customer AZJED 11012 1 2016-04-09 2016-04-23 25 Customer AZJED 10929 6 2016-03-05 2016-04-02 25 Customer AZJED 10859 1 2016-01-29 2016-02-26 26 Customer USDBG 10971 2 2016-03-24 2016-04-21 26 Customer USDBG 10860 3 2016-01-29 2016-02-26 26 Customer USDBG 10671 1 2015-09-17 2015-10-15 27 Customer WMFEA 11060 2 2016-04-30 2016-05-28 27 Customer WMFEA 11026 4 2016-04-15 2016-05-13 27 Customer WMFEA 10807 4 2015-12-31 2016-01-28 28 Customer XYUFB 10963 9 2016-03-19 2016-04-16 28 Customer XYUFB 10664 1 2015-09-10 2015-10-08 28 Customer XYUFB 10604 1 2015-07-18 2015-08-15 29 Customer MDLWA 10928 1 2016-03-05 2016-04-02 29 Customer MDLWA 10887 8 2016-02-13 2016-03-13 29 Customer MDLWA 10568 3 2015-06-13 2015-07-11 30 Customer KSLQF 